# Описание проекта

Интернет-магазин запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других.

**Задача:** поиск токсичных коментариев для отправки на модерацию.

**Цель:** обучить модель классифицировать комментарии на позитивные и негативные на наборе данных с разметкой о токсичности. Метрика f1 модели должна иметь значение не меньше 0,75.

In [72]:
import pandas as pd
import numpy as np
import re
import spacy
from scipy.sparse import vstack
from sklearn.metrics import f1_score
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from tqdm._tqdm_notebook import tqdm_notebook
from catboost import CatBoostClassifier
tqdm_notebook.pandas()

import warnings
warnings.filterwarnings(action='ignore')

# 1 Загрузка и подготовка данных

## 1.1 Загрузка данных

In [2]:
data = pd.read_csv('datasets/toxic_comments.csv')

In [3]:
data.head()

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
text     159571 non-null object
toxic    159571 non-null int64
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


Данные загружены, пропусков нет. Требуется обработка текстов.

## 1.2 Преобразование текста

**Лемматизация и замена знаков**

Прежде всего нужно привести слова к их основной форме. Используем библиотеку spacy для ускорения процесса. Для замены не буквенных символов воспользуемся регулярными выражениями

In [5]:
nlp = spacy.load('en_core_web_sm')

# определим функцию для лемматизации текстов
def lemmatize_text(row, nlp=nlp):
    doc = nlp(row)
    
    # лемматизируем
    lemma = ' '.join([token.lemma_ for token in doc]).lower()
    
    # убираем знаки, лишние пробелы и возвращаем результат
    return ' '.join(re.sub(r'[^a-zA-Z\' ]', ' ', lemma).split())

Применим функцию ко всем строкам в data['text'], проконтролируем прогресс с помощью tqdm. Так как процесс лемматизации очень длительный - применим следующую конструкцию, которая позволит не проводить лемматизацию при каждом перезапуске jupyter.

In [6]:
try:
    data_lemm = pd.read_csv('datasets/data_lemm.csv')
    print('Данные загружены!')
    del data
except FileNotFoundError:
    print('Создадим и сохраним лемматизированные тексты в файл:')
    data_lemm = data.copy()
    del data
    data_lemm['text'] = data_lemm['text'].progress_apply(lemmatize_text)   # отражаем прогресс-бар при выполнении apply
    data_lemm.to_csv('datasets/data_lemm.csv', index=False)
    print('Данные сохранены.')

Данные загружены!


In [7]:
data_lemm.head()

toxic                                               text
0      0  explanation why the edit make under pron usern...
1      0  d'aww pron match this background colour pron b...
2      0  hey man pron be really not try to edit war pro...
3      0  more pron can not make any real suggestion on ...
4      0  pron sir be pron hero any chance pron remember...

**TF-IDF для корпуса текстов**

Подгрузим стоп-слова:

In [8]:
stopwords = spacy.lang.en.stop_words.STOP_WORDS
print(stopwords)

{'n‘t', 'forty', '‘m', 'well', 'just', 'around', 'in', "'s", 'until', 'whereafter', 'such', 'among', 'does', 'sometime', 'himself', 'his', 'towards', 'meanwhile', 'against', 'mostly', 'hereby', 'fifteen', 'unless', 'few', 'ours', 'really', 'something', 'most', 'or', 'they', 'bottom', 'of', 'sixty', 'six', 'and', 'on', 'none', 'whenever', '‘d', 'why', '’re', 'him', 'never', 'whom', '’d', 'perhaps', 'might', 'almost', '’s', 'nothing', 'see', 'this', 'eleven', 'nobody', 'became', 'my', 'was', 'how', 'whereby', 'less', 'than', 'beside', 'third', 'else', 'where', 'did', 'during', 'their', 'i', 're', "'ll", 'any', 'below', 'alone', 'themselves', 'them', 'out', 'anyhow', 'anyone', 'except', 'yourself', 'two', 'anyway', 'namely', 'if', 'first', 'cannot', 'made', 'whence', 'beforehand', 'whatever', 'now', 'although', 'top', 'move', 'besides', 'been', 'everyone', 'side', 'therefore', 'four', 'either', 'take', 'down', 'are', 'everything', 'without', 'otherwise', 'thence', 'about', 'give', 'you', 

Создадим счетчик-преобразователь в вектора слов с учетом стоп-слов:

In [9]:
count_tf_idf = TfidfVectorizer(stop_words=stopwords)

Для дальнейших преобразований нужно для начала разбить признаки на обучающие и тестовые. Валидацию модели будем проводить на кроссвалидации.

In [10]:
X_train, X_test, y_train, y_test = train_test_split(data_lemm.drop('toxic', axis=1), data_lemm['toxic'], random_state=42, test_size=0.25)

Создадим вектора из набора текстов для обучающей и тестовой выборки:

In [11]:
X_train_vec = count_tf_idf.fit_transform(X_train.values.astype('U').ravel())

In [12]:
X_train_vec.shape

(119678, 133016)

На основе обученного счетчика создадим признаки для тестовых данных

In [13]:
X_test_vec = count_tf_idf.transform(X_test.values.astype('U').ravel())

In [63]:
X_test_vec.shape

(39893, 133016)

### Вывод:

1. Для набора данных была проведена лемматизация текстов с помощью библиотеки spacy. Эта библиотека выигрывает в скорости у pymystem3 и имеет все необходимое для решения задачи.
2. Произведена очистка текстов с помощью регулярных выражений.
2. Датасет разделен на обучающую и тестовую выборки, проведена процедура векторизации признаков TF-IDF, для полученных наборов и размеров выборок количество признаков равно 133016. При изменении размеров тестовой выборки это значение будет меняться.

# 2 Обучение моделей

Для выполнения задачи классификации используем модели логистической регрессии, бустер Catboost и случайную модель, на которой проверим адекватность выбраной модели.

## 2.1 Обучения моделей с гиперпараметрами по умолчанию

In [15]:
# создаем объект случайной модели
model_dummy = DummyClassifier()

# создаем объект модели логистической регрессии
model_log = LogisticRegression(random_state=42)

# создаем модель Catboost (выберем число итераций = 10 для ускорения процесса кроссвалидации)
model_cat = CatBoostClassifier(random_state=42, iterations=10)

Напишем функцию, с помощью которой можно будт оценивать и выводить на экран показатели модели:

In [16]:
def train_model(model, name):
    print('*' * 45)
    print(f'Классификация текстов для модели {name}')
    score = cross_val_score(model, X_train_vec, y_train, cv=3, scoring='f1')
    print(f'Средняя оценка метрики f1 на кроссвалидации => {score.mean()}')
    print()
    print('*' * 45)

In [17]:
for i in ((model_dummy, 'Случайная модель'), (model_log, 'Логистическая регрессия'), (model_cat, 'Catboost')):
    train_model(i[0], i[1])

*********************************************
Классификация текстов для модели Случайная модель
Средняя оценка метрики f1 на кроссвалидации => 0.10299318699638256

*********************************************
*********************************************
Классификация текстов для модели Логистическая регрессия
Средняя оценка метрики f1 на кроссвалидации => 0.7024755655896494

*********************************************
*********************************************
Классификация текстов для модели Catboost
Learning rate set to 0.5
0:	learn: 0.3442647	total: 1.21s	remaining: 10.9s
1:	learn: 0.2572535	total: 2.33s	remaining: 9.34s
2:	learn: 0.2314581	total: 3.43s	remaining: 8.01s
3:	learn: 0.2164409	total: 4.51s	remaining: 6.76s
4:	learn: 0.2081470	total: 5.61s	remaining: 5.61s
5:	learn: 0.2015730	total: 6.73s	remaining: 4.49s
6:	learn: 0.1965784	total: 7.84s	remaining: 3.36s
7:	learn: 0.1918447	total: 8.94s	remaining: 2.23s
8:	learn: 0.1881516	total: 10.1s	remaining: 1.12s
9:	learn: 0

### Вывод:

Лучшая модель на кроссвалидации с гиперпараметрами по умолчанию - модель Логистической регрессии (f1 = 0.7). Случайная модель очень плохая.

## 2.2 Оптимизация моделей

В исходных данных есть огромный дисбаланс классов:

In [18]:
y_train.value_counts()

0    107512
1     12166
Name: toxic, dtype: int64

Перевес в 8-9 раз. Добавим параметр class_weight в модель логистической регрессии и модель catboost:

In [19]:
model_log_2 = LogisticRegression(random_state=42, class_weight='balanced')

In [20]:
%%time
train_model(model_log_2, 'Логистическая регрессия (баланс)')

*********************************************
Классификация текстов для модели Логистическая регрессия (баланс)
Средняя оценка метрики f1 на кроссвалидации => 0.7454919368376918

*********************************************
Wall time: 5.34 s


In [21]:
model_cat_2 = CatBoostClassifier(random_state=42, iterations=10, class_weights=[1, np.ceil(y_train.value_counts()[0] / y_train.value_counts()[1])])

In [22]:
%%time
train_model(model_cat_2, 'Catboost (баланс)')

*********************************************
Классификация текстов для модели Catboost (баланс)
Learning rate set to 0.5
0:	learn: 0.5617049	total: 1.13s	remaining: 10.2s
1:	learn: 0.5115736	total: 2.25s	remaining: 9s
2:	learn: 0.4815795	total: 3.3s	remaining: 7.69s
3:	learn: 0.4648492	total: 4.36s	remaining: 6.54s
4:	learn: 0.4524839	total: 5.46s	remaining: 5.46s
5:	learn: 0.4408418	total: 6.54s	remaining: 4.36s
6:	learn: 0.4278960	total: 7.6s	remaining: 3.25s
7:	learn: 0.4148006	total: 8.67s	remaining: 2.17s
8:	learn: 0.4073013	total: 9.79s	remaining: 1.09s
9:	learn: 0.4015988	total: 10.9s	remaining: 0us
Learning rate set to 0.5
0:	learn: 0.5760993	total: 1.15s	remaining: 10.4s
1:	learn: 0.5139515	total: 2.19s	remaining: 8.77s
2:	learn: 0.4861711	total: 3.32s	remaining: 7.74s
3:	learn: 0.4684976	total: 4.43s	remaining: 6.65s
4:	learn: 0.4540945	total: 5.55s	remaining: 5.55s
5:	learn: 0.4401429	total: 6.66s	remaining: 4.44s
6:	learn: 0.4308809	total: 7.83s	remaining: 3.36s
7:	learn: 

Модель логистической регрессии лучше, но не достаточно (по крайней мере на кросс-валидации). Попробуем поперебирать гиперпараметры обоих моделей.

## 2.3 Подбор гиперпараметров

In [23]:
params_log = {
    'penalty': ['l1', 'l2'],
    'C': range(4, 6)
}

params_cat = {
    'iterations': [500, 1000]
}

Создадим объекты решетчатого поиска:

In [24]:
grid_log = GridSearchCV(model_log_2, params_log, cv=3, verbose=5, scoring='f1')

In [25]:
grid_cat = GridSearchCV(model_cat_2, params_cat, cv=3, verbose=5, scoring='f1')

Обучим модель логистической регрессии:

In [26]:
grid_log.fit(X_train_vec, y_train)

Fitting 3 folds for each of 4 candidates, totalling 12 fits
[CV] C=4, penalty=l1 .................................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ..................... C=4, penalty=l1, score=0.753, total=   2.2s
[CV] C=4, penalty=l1 .................................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    2.1s remaining:    0.0s


[CV] ..................... C=4, penalty=l1, score=0.769, total=   2.3s
[CV] C=4, penalty=l1 .................................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    4.4s remaining:    0.0s


[CV] ..................... C=4, penalty=l1, score=0.759, total=   2.2s
[CV] C=4, penalty=l2 .................................................


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    6.6s remaining:    0.0s


[CV] ..................... C=4, penalty=l2, score=0.754, total=   2.4s
[CV] C=4, penalty=l2 .................................................


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    9.0s remaining:    0.0s


[CV] ..................... C=4, penalty=l2, score=0.764, total=   2.9s
[CV] C=4, penalty=l2 .................................................
[CV] ..................... C=4, penalty=l2, score=0.761, total=   2.9s
[CV] C=5, penalty=l1 .................................................
[CV] ..................... C=5, penalty=l1, score=0.753, total=   2.7s
[CV] C=5, penalty=l1 .................................................
[CV] ..................... C=5, penalty=l1, score=0.768, total=   2.2s
[CV] C=5, penalty=l1 .................................................
[CV] ..................... C=5, penalty=l1, score=0.755, total=   2.3s
[CV] C=5, penalty=l2 .................................................
[CV] ..................... C=5, penalty=l2, score=0.753, total=   2.7s
[CV] C=5, penalty=l2 .................................................
[CV] ..................... C=5, penalty=l2, score=0.765, total=   2.8s
[CV] C=5, penalty=l2 .................................................
[CV] .

[Parallel(n_jobs=1)]: Done  12 out of  12 | elapsed:   30.1s finished


GridSearchCV(cv=3, error_score='raise-deprecating',
             estimator=LogisticRegression(C=1.0, class_weight='balanced',
                                          dual=False, fit_intercept=True,
                                          intercept_scaling=1, l1_ratio=None,
                                          max_iter=100, multi_class='warn',
                                          n_jobs=None, penalty='l2',
                                          random_state=42, solver='warn',
                                          tol=0.0001, verbose=0,
                                          warm_start=False),
             iid='warn', n_jobs=None,
             param_grid={'C': range(4, 6), 'penalty': ['l1', 'l2']},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring='f1', verbose=5)

In [27]:
print(f'Лучшие параметры логистической регрессии - {grid_log.best_params_}, f1_score = {grid_log.best_score_}')

Лучшие параметры логистической регрессии - {'C': 4, 'penalty': 'l1'}, f1_score = 0.7601577419506556


In [28]:
grid_cat.fit(X_train_vec, y_train)

Fitting 3 folds for each of 2 candidates, totalling 6 fits
[CV] iterations=500 ..................................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Learning rate set to 0.126211
0:	learn: 0.6362751	total: 1.15s	remaining: 9m 32s
1:	learn: 0.6016549	total: 2.45s	remaining: 10m 9s
2:	learn: 0.5820777	total: 3.65s	remaining: 10m 4s
3:	learn: 0.5640260	total: 4.76s	remaining: 9m 50s
4:	learn: 0.5491668	total: 5.82s	remaining: 9m 36s
5:	learn: 0.5377342	total: 6.95s	remaining: 9m 31s
6:	learn: 0.5269031	total: 7.99s	remaining: 9m 22s
7:	learn: 0.5189855	total: 9.09s	remaining: 9m 19s
8:	learn: 0.5123693	total: 10.2s	remaining: 9m 15s
9:	learn: 0.5077169	total: 11.3s	remaining: 9m 14s
10:	learn: 0.5025124	total: 12.4s	remaining: 9m 12s
11:	learn: 0.4951054	total: 13.5s	remaining: 9m 8s
12:	learn: 0.4904826	total: 14.6s	remaining: 9m 6s
13:	learn: 0.4827999	total: 15.7s	remaining: 9m 5s
14:	learn: 0.4760212	total: 16.8s	remaining: 9m 4s
15:	learn: 0.4699841	total: 17.9s	remaining: 9m 2s
16:	learn: 0.4666887	total: 19s	remaining: 9m
17:	learn: 0.4619926	total: 20.2s	remaining: 8m 59s
18:	learn: 0.4583475	total: 21.2s	remaining: 8m 57s
19:

156:	learn: 0.2995849	total: 2m 55s	remaining: 6m 22s
157:	learn: 0.2990171	total: 2m 56s	remaining: 6m 21s
158:	learn: 0.2984864	total: 2m 57s	remaining: 6m 20s
159:	learn: 0.2979781	total: 2m 58s	remaining: 6m 19s
160:	learn: 0.2972459	total: 2m 59s	remaining: 6m 17s
161:	learn: 0.2966701	total: 3m	remaining: 6m 16s
162:	learn: 0.2954423	total: 3m 1s	remaining: 6m 15s
163:	learn: 0.2948613	total: 3m 2s	remaining: 6m 14s
164:	learn: 0.2944928	total: 3m 3s	remaining: 6m 13s
165:	learn: 0.2940773	total: 3m 5s	remaining: 6m 12s
166:	learn: 0.2934495	total: 3m 6s	remaining: 6m 11s
167:	learn: 0.2929685	total: 3m 7s	remaining: 6m 10s
168:	learn: 0.2925516	total: 3m 8s	remaining: 6m 8s
169:	learn: 0.2920343	total: 3m 9s	remaining: 6m 7s
170:	learn: 0.2916409	total: 3m 10s	remaining: 6m 6s
171:	learn: 0.2909926	total: 3m 11s	remaining: 6m 5s
172:	learn: 0.2904815	total: 3m 13s	remaining: 6m 4s
173:	learn: 0.2900400	total: 3m 14s	remaining: 6m 4s
174:	learn: 0.2896326	total: 3m 15s	remaining:

309:	learn: 0.2406652	total: 5m 41s	remaining: 3m 29s
310:	learn: 0.2403833	total: 5m 42s	remaining: 3m 28s
311:	learn: 0.2402093	total: 5m 43s	remaining: 3m 26s
312:	learn: 0.2399553	total: 5m 44s	remaining: 3m 25s
313:	learn: 0.2394302	total: 5m 45s	remaining: 3m 24s
314:	learn: 0.2391364	total: 5m 46s	remaining: 3m 23s
315:	learn: 0.2388813	total: 5m 47s	remaining: 3m 22s
316:	learn: 0.2385821	total: 5m 48s	remaining: 3m 21s
317:	learn: 0.2382709	total: 5m 49s	remaining: 3m 20s
318:	learn: 0.2380635	total: 5m 50s	remaining: 3m 19s
319:	learn: 0.2378327	total: 5m 52s	remaining: 3m 18s
320:	learn: 0.2373964	total: 5m 53s	remaining: 3m 17s
321:	learn: 0.2371836	total: 5m 54s	remaining: 3m 15s
322:	learn: 0.2370510	total: 5m 55s	remaining: 3m 14s
323:	learn: 0.2368608	total: 5m 56s	remaining: 3m 13s
324:	learn: 0.2366144	total: 5m 57s	remaining: 3m 12s
325:	learn: 0.2364876	total: 5m 59s	remaining: 3m 11s
326:	learn: 0.2363129	total: 6m	remaining: 3m 10s
327:	learn: 0.2360153	total: 6m 

463:	learn: 0.2077432	total: 8m 26s	remaining: 39.3s
464:	learn: 0.2076293	total: 8m 27s	remaining: 38.2s
465:	learn: 0.2073771	total: 8m 29s	remaining: 37.1s
466:	learn: 0.2072844	total: 8m 30s	remaining: 36.1s
467:	learn: 0.2071146	total: 8m 31s	remaining: 35s
468:	learn: 0.2070216	total: 8m 32s	remaining: 33.9s
469:	learn: 0.2069295	total: 8m 33s	remaining: 32.8s
470:	learn: 0.2066082	total: 8m 34s	remaining: 31.7s
471:	learn: 0.2064567	total: 8m 35s	remaining: 30.6s
472:	learn: 0.2063676	total: 8m 36s	remaining: 29.5s
473:	learn: 0.2061621	total: 8m 37s	remaining: 28.4s
474:	learn: 0.2059452	total: 8m 38s	remaining: 27.3s
475:	learn: 0.2056842	total: 8m 39s	remaining: 26.2s
476:	learn: 0.2055932	total: 8m 40s	remaining: 25.1s
477:	learn: 0.2054168	total: 8m 42s	remaining: 24s
478:	learn: 0.2052205	total: 8m 43s	remaining: 22.9s
479:	learn: 0.2050941	total: 8m 44s	remaining: 21.8s
480:	learn: 0.2048722	total: 8m 45s	remaining: 20.7s
481:	learn: 0.2046794	total: 8m 46s	remaining: 19.

[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:  9.3min remaining:    0.0s


Learning rate set to 0.126212
0:	learn: 0.6445913	total: 1.18s	remaining: 9m 47s
1:	learn: 0.6068363	total: 2.26s	remaining: 9m 23s
2:	learn: 0.5834099	total: 3.36s	remaining: 9m 16s
3:	learn: 0.5633859	total: 4.47s	remaining: 9m 14s
4:	learn: 0.5511748	total: 5.57s	remaining: 9m 11s
5:	learn: 0.5387032	total: 6.71s	remaining: 9m 12s
6:	learn: 0.5301523	total: 7.82s	remaining: 9m 10s
7:	learn: 0.5206508	total: 8.98s	remaining: 9m 12s
8:	learn: 0.5098895	total: 10.1s	remaining: 9m 10s
9:	learn: 0.5038509	total: 11.2s	remaining: 9m 10s
10:	learn: 0.4994883	total: 12.4s	remaining: 9m 9s
11:	learn: 0.4968700	total: 13.4s	remaining: 9m 5s
12:	learn: 0.4910157	total: 14.5s	remaining: 9m 3s
13:	learn: 0.4847904	total: 15.6s	remaining: 9m
14:	learn: 0.4800528	total: 16.7s	remaining: 8m 58s
15:	learn: 0.4742670	total: 17.7s	remaining: 8m 55s
16:	learn: 0.4694428	total: 18.8s	remaining: 8m 53s
17:	learn: 0.4646512	total: 19.9s	remaining: 8m 51s
18:	learn: 0.4611951	total: 20.9s	remaining: 8m 50s

156:	learn: 0.3053281	total: 2m 51s	remaining: 6m 14s
157:	learn: 0.3041434	total: 2m 52s	remaining: 6m 13s
158:	learn: 0.3035513	total: 2m 53s	remaining: 6m 12s
159:	learn: 0.3028878	total: 2m 54s	remaining: 6m 11s
160:	learn: 0.3024635	total: 2m 55s	remaining: 6m 10s
161:	learn: 0.3019665	total: 2m 56s	remaining: 6m 9s
162:	learn: 0.3014091	total: 2m 57s	remaining: 6m 7s
163:	learn: 0.3008788	total: 2m 58s	remaining: 6m 6s
164:	learn: 0.3004791	total: 3m	remaining: 6m 5s
165:	learn: 0.2997230	total: 3m 1s	remaining: 6m 4s
166:	learn: 0.2990164	total: 3m 2s	remaining: 6m 3s
167:	learn: 0.2985567	total: 3m 3s	remaining: 6m 2s
168:	learn: 0.2979845	total: 3m 4s	remaining: 6m 1s
169:	learn: 0.2975509	total: 3m 5s	remaining: 6m
170:	learn: 0.2970632	total: 3m 6s	remaining: 5m 58s
171:	learn: 0.2965281	total: 3m 7s	remaining: 5m 57s
172:	learn: 0.2960944	total: 3m 8s	remaining: 5m 56s
173:	learn: 0.2956394	total: 3m 9s	remaining: 5m 55s
174:	learn: 0.2951911	total: 3m 10s	remaining: 5m 54s

310:	learn: 0.2461074	total: 5m 39s	remaining: 3m 26s
311:	learn: 0.2458571	total: 5m 40s	remaining: 3m 25s
312:	learn: 0.2455800	total: 5m 42s	remaining: 3m 24s
313:	learn: 0.2453118	total: 5m 43s	remaining: 3m 23s
314:	learn: 0.2451005	total: 5m 44s	remaining: 3m 22s
315:	learn: 0.2447638	total: 5m 45s	remaining: 3m 21s
316:	learn: 0.2443341	total: 5m 46s	remaining: 3m 19s
317:	learn: 0.2440721	total: 5m 47s	remaining: 3m 18s
318:	learn: 0.2439404	total: 5m 48s	remaining: 3m 17s
319:	learn: 0.2435924	total: 5m 49s	remaining: 3m 16s
320:	learn: 0.2434655	total: 5m 50s	remaining: 3m 15s
321:	learn: 0.2431574	total: 5m 51s	remaining: 3m 14s
322:	learn: 0.2428968	total: 5m 52s	remaining: 3m 13s
323:	learn: 0.2427406	total: 5m 53s	remaining: 3m 12s
324:	learn: 0.2424947	total: 5m 55s	remaining: 3m 11s
325:	learn: 0.2422662	total: 5m 56s	remaining: 3m 10s
326:	learn: 0.2419250	total: 5m 57s	remaining: 3m 9s
327:	learn: 0.2416102	total: 5m 59s	remaining: 3m 8s
328:	learn: 0.2413235	total: 6

464:	learn: 0.2121555	total: 8m 32s	remaining: 38.6s
465:	learn: 0.2119493	total: 8m 33s	remaining: 37.5s
466:	learn: 0.2118550	total: 8m 34s	remaining: 36.4s
467:	learn: 0.2116332	total: 8m 35s	remaining: 35.3s
468:	learn: 0.2115044	total: 8m 36s	remaining: 34.1s
469:	learn: 0.2114120	total: 8m 37s	remaining: 33s
470:	learn: 0.2113187	total: 8m 38s	remaining: 31.9s
471:	learn: 0.2112255	total: 8m 40s	remaining: 30.8s
472:	learn: 0.2110224	total: 8m 41s	remaining: 29.7s
473:	learn: 0.2107896	total: 8m 42s	remaining: 28.6s
474:	learn: 0.2106991	total: 8m 43s	remaining: 27.5s
475:	learn: 0.2103152	total: 8m 44s	remaining: 26.4s
476:	learn: 0.2101408	total: 8m 45s	remaining: 25.3s
477:	learn: 0.2100482	total: 8m 46s	remaining: 24.2s
478:	learn: 0.2098139	total: 8m 47s	remaining: 23.1s
479:	learn: 0.2094899	total: 8m 48s	remaining: 22s
480:	learn: 0.2093031	total: 8m 49s	remaining: 20.9s
481:	learn: 0.2090951	total: 8m 51s	remaining: 19.8s
482:	learn: 0.2090045	total: 8m 52s	remaining: 18.

[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed: 18.7min remaining:    0.0s


Learning rate set to 0.126212
0:	learn: 0.6384723	total: 1.17s	remaining: 9m 45s
1:	learn: 0.6059812	total: 2.49s	remaining: 10m 20s
2:	learn: 0.5812950	total: 4.26s	remaining: 11m 45s
3:	learn: 0.5622380	total: 5.53s	remaining: 11m 25s
4:	learn: 0.5469124	total: 6.76s	remaining: 11m 8s
5:	learn: 0.5353417	total: 7.95s	remaining: 10m 54s
6:	learn: 0.5293739	total: 9.18s	remaining: 10m 46s
7:	learn: 0.5203800	total: 10.3s	remaining: 10m 34s
8:	learn: 0.5110376	total: 11.5s	remaining: 10m 25s
9:	learn: 0.5042504	total: 12.6s	remaining: 10m 18s
10:	learn: 0.4982406	total: 13.9s	remaining: 10m 17s
11:	learn: 0.4939358	total: 15.1s	remaining: 10m 12s
12:	learn: 0.4875872	total: 16.2s	remaining: 10m 7s
13:	learn: 0.4824898	total: 17.4s	remaining: 10m 3s
14:	learn: 0.4774037	total: 18.5s	remaining: 9m 59s
15:	learn: 0.4734526	total: 19.7s	remaining: 9m 56s
16:	learn: 0.4699076	total: 20.8s	remaining: 9m 52s
17:	learn: 0.4651916	total: 21.9s	remaining: 9m 47s
18:	learn: 0.4624064	total: 23.1s	

155:	learn: 0.3031802	total: 2m 54s	remaining: 6m 24s
156:	learn: 0.3027530	total: 2m 55s	remaining: 6m 23s
157:	learn: 0.3016842	total: 2m 56s	remaining: 6m 22s
158:	learn: 0.3011364	total: 2m 57s	remaining: 6m 20s
159:	learn: 0.3005207	total: 2m 58s	remaining: 6m 19s
160:	learn: 0.3000926	total: 2m 59s	remaining: 6m 18s
161:	learn: 0.2990525	total: 3m	remaining: 6m 17s
162:	learn: 0.2985582	total: 3m 1s	remaining: 6m 16s
163:	learn: 0.2980697	total: 3m 2s	remaining: 6m 14s
164:	learn: 0.2975960	total: 3m 4s	remaining: 6m 13s
165:	learn: 0.2971875	total: 3m 5s	remaining: 6m 12s
166:	learn: 0.2966133	total: 3m 6s	remaining: 6m 11s
167:	learn: 0.2961028	total: 3m 7s	remaining: 6m 10s
168:	learn: 0.2957827	total: 3m 8s	remaining: 6m 9s
169:	learn: 0.2952601	total: 3m 9s	remaining: 6m 7s
170:	learn: 0.2948006	total: 3m 10s	remaining: 6m 6s
171:	learn: 0.2940618	total: 3m 11s	remaining: 6m 5s
172:	learn: 0.2933606	total: 3m 12s	remaining: 6m 4s
173:	learn: 0.2928500	total: 3m 13s	remaining

308:	learn: 0.2438062	total: 5m 40s	remaining: 3m 30s
309:	learn: 0.2434505	total: 5m 42s	remaining: 3m 29s
310:	learn: 0.2432434	total: 5m 43s	remaining: 3m 28s
311:	learn: 0.2429695	total: 5m 44s	remaining: 3m 27s
312:	learn: 0.2426919	total: 5m 45s	remaining: 3m 26s
313:	learn: 0.2424471	total: 5m 46s	remaining: 3m 25s
314:	learn: 0.2423138	total: 5m 47s	remaining: 3m 24s
315:	learn: 0.2418840	total: 5m 48s	remaining: 3m 22s
316:	learn: 0.2417268	total: 5m 49s	remaining: 3m 21s
317:	learn: 0.2414861	total: 5m 50s	remaining: 3m 20s
318:	learn: 0.2411783	total: 5m 51s	remaining: 3m 19s
319:	learn: 0.2408678	total: 5m 52s	remaining: 3m 18s
320:	learn: 0.2405446	total: 5m 54s	remaining: 3m 17s
321:	learn: 0.2403079	total: 5m 55s	remaining: 3m 16s
322:	learn: 0.2400189	total: 5m 56s	remaining: 3m 15s
323:	learn: 0.2396986	total: 5m 57s	remaining: 3m 14s
324:	learn: 0.2393398	total: 5m 58s	remaining: 3m 12s
325:	learn: 0.2390655	total: 5m 59s	remaining: 3m 11s
326:	learn: 0.2388825	total:

462:	learn: 0.2103361	total: 8m 31s	remaining: 40.9s
463:	learn: 0.2101515	total: 8m 32s	remaining: 39.8s
464:	learn: 0.2099509	total: 8m 33s	remaining: 38.7s
465:	learn: 0.2097437	total: 8m 34s	remaining: 37.6s
466:	learn: 0.2096515	total: 8m 35s	remaining: 36.5s
467:	learn: 0.2095585	total: 8m 37s	remaining: 35.4s
468:	learn: 0.2094657	total: 8m 38s	remaining: 34.3s
469:	learn: 0.2093756	total: 8m 39s	remaining: 33.1s
470:	learn: 0.2092278	total: 8m 40s	remaining: 32s
471:	learn: 0.2090333	total: 8m 41s	remaining: 30.9s
472:	learn: 0.2088908	total: 8m 42s	remaining: 29.8s
473:	learn: 0.2087132	total: 8m 43s	remaining: 28.7s
474:	learn: 0.2086259	total: 8m 44s	remaining: 27.6s
475:	learn: 0.2084565	total: 8m 45s	remaining: 26.5s
476:	learn: 0.2080448	total: 8m 47s	remaining: 25.4s
477:	learn: 0.2077943	total: 8m 48s	remaining: 24.3s
478:	learn: 0.2075990	total: 8m 49s	remaining: 23.2s
479:	learn: 0.2074796	total: 8m 50s	remaining: 22.1s
480:	learn: 0.2072821	total: 8m 51s	remaining: 2

[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed: 28.0min remaining:    0.0s


Learning rate set to 0.066842
0:	learn: 0.6603498	total: 1.25s	remaining: 20m 45s
1:	learn: 0.6360413	total: 2.34s	remaining: 19m 27s
2:	learn: 0.6155865	total: 3.42s	remaining: 18m 55s
3:	learn: 0.6014322	total: 4.48s	remaining: 18m 36s
4:	learn: 0.5899260	total: 5.6s	remaining: 18m 34s
5:	learn: 0.5781112	total: 6.8s	remaining: 18m 46s
6:	learn: 0.5688896	total: 7.87s	remaining: 18m 36s
7:	learn: 0.5600868	total: 8.98s	remaining: 18m 33s
8:	learn: 0.5539731	total: 10.3s	remaining: 18m 50s
9:	learn: 0.5493491	total: 11.4s	remaining: 18m 46s
10:	learn: 0.5460095	total: 12.6s	remaining: 18m 51s
11:	learn: 0.5385647	total: 13.7s	remaining: 18m 49s
12:	learn: 0.5336710	total: 14.8s	remaining: 18m 46s
13:	learn: 0.5285547	total: 15.9s	remaining: 18m 41s
14:	learn: 0.5220158	total: 17s	remaining: 18m 36s
15:	learn: 0.5162347	total: 18.3s	remaining: 18m 43s
16:	learn: 0.5131249	total: 19.5s	remaining: 18m 45s
17:	learn: 0.5087730	total: 20.6s	remaining: 18m 46s
18:	learn: 0.5055183	total: 21

153:	learn: 0.3611734	total: 2m 50s	remaining: 15m 34s
154:	learn: 0.3605192	total: 2m 51s	remaining: 15m 33s
155:	learn: 0.3600552	total: 2m 52s	remaining: 15m 33s
156:	learn: 0.3597397	total: 2m 53s	remaining: 15m 32s
157:	learn: 0.3592117	total: 2m 54s	remaining: 15m 30s
158:	learn: 0.3587443	total: 2m 55s	remaining: 15m 29s
159:	learn: 0.3579979	total: 2m 56s	remaining: 15m 29s
160:	learn: 0.3576032	total: 2m 58s	remaining: 15m 28s
161:	learn: 0.3572134	total: 2m 59s	remaining: 15m 27s
162:	learn: 0.3568467	total: 3m	remaining: 15m 25s
163:	learn: 0.3564178	total: 3m 1s	remaining: 15m 24s
164:	learn: 0.3559109	total: 3m 2s	remaining: 15m 23s
165:	learn: 0.3555218	total: 3m 3s	remaining: 15m 22s
166:	learn: 0.3550472	total: 3m 4s	remaining: 15m 20s
167:	learn: 0.3546994	total: 3m 5s	remaining: 15m 19s
168:	learn: 0.3538146	total: 3m 6s	remaining: 15m 18s
169:	learn: 0.3531043	total: 3m 7s	remaining: 15m 17s
170:	learn: 0.3526585	total: 3m 8s	remaining: 15m 16s
171:	learn: 0.3519875	

304:	learn: 0.2972095	total: 5m 34s	remaining: 12m 42s
305:	learn: 0.2969725	total: 5m 35s	remaining: 12m 41s
306:	learn: 0.2967354	total: 5m 36s	remaining: 12m 40s
307:	learn: 0.2964633	total: 5m 37s	remaining: 12m 38s
308:	learn: 0.2962302	total: 5m 38s	remaining: 12m 37s
309:	learn: 0.2960101	total: 5m 39s	remaining: 12m 36s
310:	learn: 0.2958003	total: 5m 40s	remaining: 12m 35s
311:	learn: 0.2954785	total: 5m 42s	remaining: 12m 34s
312:	learn: 0.2952038	total: 5m 43s	remaining: 12m 33s
313:	learn: 0.2949734	total: 5m 44s	remaining: 12m 31s
314:	learn: 0.2947477	total: 5m 45s	remaining: 12m 30s
315:	learn: 0.2944350	total: 5m 46s	remaining: 12m 29s
316:	learn: 0.2941017	total: 5m 47s	remaining: 12m 28s
317:	learn: 0.2939106	total: 5m 48s	remaining: 12m 27s
318:	learn: 0.2934296	total: 5m 49s	remaining: 12m 26s
319:	learn: 0.2931782	total: 5m 50s	remaining: 12m 25s
320:	learn: 0.2928969	total: 5m 51s	remaining: 12m 24s
321:	learn: 0.2925850	total: 5m 53s	remaining: 12m 23s
322:	learn

455:	learn: 0.2621398	total: 8m 23s	remaining: 10m
456:	learn: 0.2618846	total: 8m 24s	remaining: 9m 59s
457:	learn: 0.2617090	total: 8m 25s	remaining: 9m 58s
458:	learn: 0.2615493	total: 8m 26s	remaining: 9m 57s
459:	learn: 0.2613701	total: 8m 27s	remaining: 9m 56s
460:	learn: 0.2611445	total: 8m 29s	remaining: 9m 55s
461:	learn: 0.2610329	total: 8m 30s	remaining: 9m 54s
462:	learn: 0.2608606	total: 8m 31s	remaining: 9m 53s
463:	learn: 0.2606497	total: 8m 32s	remaining: 9m 52s
464:	learn: 0.2604746	total: 8m 34s	remaining: 9m 51s
465:	learn: 0.2602605	total: 8m 35s	remaining: 9m 50s
466:	learn: 0.2601244	total: 8m 36s	remaining: 9m 49s
467:	learn: 0.2599017	total: 8m 37s	remaining: 9m 48s
468:	learn: 0.2597490	total: 8m 38s	remaining: 9m 47s
469:	learn: 0.2595548	total: 8m 39s	remaining: 9m 46s
470:	learn: 0.2593566	total: 8m 41s	remaining: 9m 45s
471:	learn: 0.2590965	total: 8m 42s	remaining: 9m 44s
472:	learn: 0.2589879	total: 8m 44s	remaining: 9m 43s
473:	learn: 0.2588005	total: 8m

607:	learn: 0.2381926	total: 11m 14s	remaining: 7m 15s
608:	learn: 0.2379990	total: 11m 15s	remaining: 7m 14s
609:	learn: 0.2379121	total: 11m 17s	remaining: 7m 12s
610:	learn: 0.2377169	total: 11m 18s	remaining: 7m 11s
611:	learn: 0.2373525	total: 11m 19s	remaining: 7m 10s
612:	learn: 0.2372053	total: 11m 20s	remaining: 7m 9s
613:	learn: 0.2370436	total: 11m 21s	remaining: 7m 8s
614:	learn: 0.2369624	total: 11m 22s	remaining: 7m 7s
615:	learn: 0.2368082	total: 11m 23s	remaining: 7m 6s
616:	learn: 0.2367408	total: 11m 24s	remaining: 7m 4s
617:	learn: 0.2365799	total: 11m 25s	remaining: 7m 3s
618:	learn: 0.2364170	total: 11m 26s	remaining: 7m 2s
619:	learn: 0.2362544	total: 11m 27s	remaining: 7m 1s
620:	learn: 0.2361423	total: 11m 28s	remaining: 7m
621:	learn: 0.2360291	total: 11m 30s	remaining: 6m 59s
622:	learn: 0.2358613	total: 11m 31s	remaining: 6m 58s
623:	learn: 0.2357349	total: 11m 32s	remaining: 6m 57s
624:	learn: 0.2355926	total: 11m 33s	remaining: 6m 56s
625:	learn: 0.2355304	

758:	learn: 0.2194299	total: 13m 59s	remaining: 4m 26s
759:	learn: 0.2193763	total: 14m	remaining: 4m 25s
760:	learn: 0.2190935	total: 14m 1s	remaining: 4m 24s
761:	learn: 0.2189925	total: 14m 2s	remaining: 4m 23s
762:	learn: 0.2188789	total: 14m 3s	remaining: 4m 21s
763:	learn: 0.2187583	total: 14m 4s	remaining: 4m 20s
764:	learn: 0.2185825	total: 14m 5s	remaining: 4m 19s
765:	learn: 0.2184765	total: 14m 6s	remaining: 4m 18s
766:	learn: 0.2183782	total: 14m 7s	remaining: 4m 17s
767:	learn: 0.2182622	total: 14m 8s	remaining: 4m 16s
768:	learn: 0.2181800	total: 14m 9s	remaining: 4m 15s
769:	learn: 0.2181276	total: 14m 10s	remaining: 4m 14s
770:	learn: 0.2180068	total: 14m 11s	remaining: 4m 12s
771:	learn: 0.2179019	total: 14m 12s	remaining: 4m 11s
772:	learn: 0.2177869	total: 14m 13s	remaining: 4m 10s
773:	learn: 0.2176695	total: 14m 14s	remaining: 4m 9s
774:	learn: 0.2174200	total: 14m 16s	remaining: 4m 8s
775:	learn: 0.2172489	total: 14m 17s	remaining: 4m 7s
776:	learn: 0.2171253	tota

909:	learn: 0.2048214	total: 16m 42s	remaining: 1m 39s
910:	learn: 0.2047072	total: 16m 43s	remaining: 1m 38s
911:	learn: 0.2046507	total: 16m 44s	remaining: 1m 36s
912:	learn: 0.2044627	total: 16m 45s	remaining: 1m 35s
913:	learn: 0.2043631	total: 16m 46s	remaining: 1m 34s
914:	learn: 0.2042714	total: 16m 47s	remaining: 1m 33s
915:	learn: 0.2042262	total: 16m 48s	remaining: 1m 32s
916:	learn: 0.2041312	total: 16m 49s	remaining: 1m 31s
917:	learn: 0.2040368	total: 16m 50s	remaining: 1m 30s
918:	learn: 0.2039218	total: 16m 51s	remaining: 1m 29s
919:	learn: 0.2037974	total: 16m 52s	remaining: 1m 28s
920:	learn: 0.2036672	total: 16m 54s	remaining: 1m 26s
921:	learn: 0.2036098	total: 16m 55s	remaining: 1m 25s
922:	learn: 0.2035643	total: 16m 56s	remaining: 1m 24s
923:	learn: 0.2034668	total: 16m 57s	remaining: 1m 23s
924:	learn: 0.2033801	total: 16m 58s	remaining: 1m 22s
925:	learn: 0.2033358	total: 16m 59s	remaining: 1m 21s
926:	learn: 0.2031953	total: 17m	remaining: 1m 20s
927:	learn: 0.

[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed: 46.6min remaining:    0.0s


Learning rate set to 0.066843
0:	learn: 0.6653748	total: 1.13s	remaining: 18m 50s
1:	learn: 0.6398731	total: 2.2s	remaining: 18m 19s
2:	learn: 0.6178939	total: 3.3s	remaining: 18m 15s
3:	learn: 0.6031199	total: 4.56s	remaining: 18m 55s
4:	learn: 0.5939297	total: 5.79s	remaining: 19m 13s
5:	learn: 0.5808084	total: 7.08s	remaining: 19m 32s
6:	learn: 0.5706992	total: 8.52s	remaining: 20m 8s
7:	learn: 0.5649557	total: 9.88s	remaining: 20m 25s
8:	learn: 0.5567708	total: 11.1s	remaining: 20m 25s
9:	learn: 0.5482702	total: 12.6s	remaining: 20m 43s
10:	learn: 0.5428463	total: 13.9s	remaining: 20m 53s
11:	learn: 0.5380749	total: 15.3s	remaining: 20m 58s
12:	learn: 0.5332921	total: 16.7s	remaining: 21m 7s
13:	learn: 0.5289240	total: 18.1s	remaining: 21m 16s
14:	learn: 0.5240098	total: 19.8s	remaining: 21m 39s
15:	learn: 0.5187295	total: 21.3s	remaining: 21m 50s
16:	learn: 0.5156312	total: 23s	remaining: 22m 11s
17:	learn: 0.5109086	total: 24.6s	remaining: 22m 22s
18:	learn: 0.5085004	total: 26.1

153:	learn: 0.3649002	total: 3m	remaining: 16m 31s
154:	learn: 0.3642577	total: 3m 1s	remaining: 16m 29s
155:	learn: 0.3638361	total: 3m 2s	remaining: 16m 27s
156:	learn: 0.3635039	total: 3m 3s	remaining: 16m 26s
157:	learn: 0.3628772	total: 3m 4s	remaining: 16m 24s
158:	learn: 0.3623060	total: 3m 5s	remaining: 16m 23s
159:	learn: 0.3615379	total: 3m 6s	remaining: 16m 21s
160:	learn: 0.3611316	total: 3m 8s	remaining: 16m 20s
161:	learn: 0.3606503	total: 3m 9s	remaining: 16m 18s
162:	learn: 0.3602756	total: 3m 10s	remaining: 16m 17s
163:	learn: 0.3598241	total: 3m 11s	remaining: 16m 15s
164:	learn: 0.3593611	total: 3m 12s	remaining: 16m 14s
165:	learn: 0.3590006	total: 3m 13s	remaining: 16m 12s
166:	learn: 0.3586121	total: 3m 14s	remaining: 16m 12s
167:	learn: 0.3581088	total: 3m 16s	remaining: 16m 10s
168:	learn: 0.3577106	total: 3m 17s	remaining: 16m 10s
169:	learn: 0.3573952	total: 3m 18s	remaining: 16m 8s
170:	learn: 0.3568499	total: 3m 19s	remaining: 16m 7s
171:	learn: 0.3560865	to

304:	learn: 0.3029381	total: 5m 49s	remaining: 13m 16s
305:	learn: 0.3027082	total: 5m 50s	remaining: 13m 15s
306:	learn: 0.3024554	total: 5m 51s	remaining: 13m 14s
307:	learn: 0.3021785	total: 5m 52s	remaining: 13m 12s
308:	learn: 0.3019063	total: 5m 53s	remaining: 13m 11s
309:	learn: 0.3011948	total: 5m 55s	remaining: 13m 10s
310:	learn: 0.3008712	total: 5m 56s	remaining: 13m 8s
311:	learn: 0.3005232	total: 5m 57s	remaining: 13m 7s
312:	learn: 0.3002881	total: 5m 58s	remaining: 13m 6s
313:	learn: 0.3000072	total: 5m 59s	remaining: 13m 5s
314:	learn: 0.2997871	total: 6m	remaining: 13m 3s
315:	learn: 0.2995439	total: 6m 1s	remaining: 13m 2s
316:	learn: 0.2993567	total: 6m 2s	remaining: 13m 1s
317:	learn: 0.2990592	total: 6m 3s	remaining: 13m
318:	learn: 0.2987808	total: 6m 4s	remaining: 12m 58s
319:	learn: 0.2985388	total: 6m 5s	remaining: 12m 57s
320:	learn: 0.2983262	total: 6m 6s	remaining: 12m 56s
321:	learn: 0.2980818	total: 6m 8s	remaining: 12m 54s
322:	learn: 0.2978382	total: 6m 

455:	learn: 0.2670319	total: 8m 47s	remaining: 10m 28s
456:	learn: 0.2667959	total: 8m 48s	remaining: 10m 27s
457:	learn: 0.2665844	total: 8m 49s	remaining: 10m 26s
458:	learn: 0.2664782	total: 8m 50s	remaining: 10m 25s
459:	learn: 0.2663001	total: 8m 51s	remaining: 10m 23s
460:	learn: 0.2661436	total: 8m 52s	remaining: 10m 22s
461:	learn: 0.2657507	total: 8m 53s	remaining: 10m 21s
462:	learn: 0.2655954	total: 8m 54s	remaining: 10m 20s
463:	learn: 0.2654765	total: 8m 55s	remaining: 10m 18s
464:	learn: 0.2653405	total: 8m 56s	remaining: 10m 17s
465:	learn: 0.2652503	total: 8m 57s	remaining: 10m 16s
466:	learn: 0.2650546	total: 8m 58s	remaining: 10m 15s
467:	learn: 0.2648928	total: 9m	remaining: 10m 13s
468:	learn: 0.2646704	total: 9m 1s	remaining: 10m 12s
469:	learn: 0.2645420	total: 9m 2s	remaining: 10m 11s
470:	learn: 0.2643819	total: 9m 3s	remaining: 10m 10s
471:	learn: 0.2642144	total: 9m 4s	remaining: 10m 8s
472:	learn: 0.2640512	total: 9m 5s	remaining: 10m 7s
473:	learn: 0.2638895

607:	learn: 0.2426045	total: 11m 35s	remaining: 7m 28s
608:	learn: 0.2424410	total: 11m 36s	remaining: 7m 27s
609:	learn: 0.2422134	total: 11m 37s	remaining: 7m 26s
610:	learn: 0.2421460	total: 11m 39s	remaining: 7m 25s
611:	learn: 0.2420178	total: 11m 40s	remaining: 7m 23s
612:	learn: 0.2418564	total: 11m 41s	remaining: 7m 22s
613:	learn: 0.2417351	total: 11m 42s	remaining: 7m 21s
614:	learn: 0.2415882	total: 11m 43s	remaining: 7m 20s
615:	learn: 0.2415028	total: 11m 44s	remaining: 7m 19s
616:	learn: 0.2413265	total: 11m 45s	remaining: 7m 17s
617:	learn: 0.2412420	total: 11m 46s	remaining: 7m 16s
618:	learn: 0.2411005	total: 11m 47s	remaining: 7m 15s
619:	learn: 0.2409429	total: 11m 48s	remaining: 7m 14s
620:	learn: 0.2408767	total: 11m 49s	remaining: 7m 13s
621:	learn: 0.2407808	total: 11m 50s	remaining: 7m 12s
622:	learn: 0.2406989	total: 11m 52s	remaining: 7m 10s
623:	learn: 0.2405689	total: 11m 53s	remaining: 7m 9s
624:	learn: 0.2404221	total: 11m 54s	remaining: 7m 8s
625:	learn: 

758:	learn: 0.2243300	total: 14m 18s	remaining: 4m 32s
759:	learn: 0.2242074	total: 14m 19s	remaining: 4m 31s
760:	learn: 0.2240846	total: 14m 20s	remaining: 4m 30s
761:	learn: 0.2239479	total: 14m 22s	remaining: 4m 29s
762:	learn: 0.2238295	total: 14m 23s	remaining: 4m 28s
763:	learn: 0.2237574	total: 14m 24s	remaining: 4m 26s
764:	learn: 0.2236169	total: 14m 25s	remaining: 4m 25s
765:	learn: 0.2234964	total: 14m 26s	remaining: 4m 24s
766:	learn: 0.2233957	total: 14m 27s	remaining: 4m 23s
767:	learn: 0.2233415	total: 14m 28s	remaining: 4m 22s
768:	learn: 0.2232893	total: 14m 29s	remaining: 4m 21s
769:	learn: 0.2231566	total: 14m 30s	remaining: 4m 20s
770:	learn: 0.2229853	total: 14m 31s	remaining: 4m 18s
771:	learn: 0.2228678	total: 14m 32s	remaining: 4m 17s
772:	learn: 0.2227720	total: 14m 33s	remaining: 4m 16s
773:	learn: 0.2227188	total: 14m 34s	remaining: 4m 15s
774:	learn: 0.2226048	total: 14m 36s	remaining: 4m 14s
775:	learn: 0.2224794	total: 14m 37s	remaining: 4m 13s
776:	learn

909:	learn: 0.2091681	total: 17m 1s	remaining: 1m 41s
910:	learn: 0.2090605	total: 17m 2s	remaining: 1m 39s
911:	learn: 0.2089977	total: 17m 4s	remaining: 1m 38s
912:	learn: 0.2088981	total: 17m 5s	remaining: 1m 37s
913:	learn: 0.2088175	total: 17m 6s	remaining: 1m 36s
914:	learn: 0.2087130	total: 17m 7s	remaining: 1m 35s
915:	learn: 0.2086218	total: 17m 8s	remaining: 1m 34s
916:	learn: 0.2084974	total: 17m 9s	remaining: 1m 33s
917:	learn: 0.2084505	total: 17m 10s	remaining: 1m 32s
918:	learn: 0.2083519	total: 17m 11s	remaining: 1m 30s
919:	learn: 0.2082616	total: 17m 12s	remaining: 1m 29s
920:	learn: 0.2081560	total: 17m 13s	remaining: 1m 28s
921:	learn: 0.2081093	total: 17m 14s	remaining: 1m 27s
922:	learn: 0.2079036	total: 17m 15s	remaining: 1m 26s
923:	learn: 0.2078579	total: 17m 16s	remaining: 1m 25s
924:	learn: 0.2077600	total: 17m 18s	remaining: 1m 24s
925:	learn: 0.2076714	total: 17m 19s	remaining: 1m 23s
926:	learn: 0.2075621	total: 17m 20s	remaining: 1m 21s
927:	learn: 0.2074

60:	learn: 0.4295492	total: 1m 7s	remaining: 17m 22s
61:	learn: 0.4285217	total: 1m 8s	remaining: 17m 21s
62:	learn: 0.4275807	total: 1m 9s	remaining: 17m 21s
63:	learn: 0.4266897	total: 1m 11s	remaining: 17m 20s
64:	learn: 0.4252026	total: 1m 12s	remaining: 17m 19s
65:	learn: 0.4239411	total: 1m 13s	remaining: 17m 18s
66:	learn: 0.4231874	total: 1m 14s	remaining: 17m 17s
67:	learn: 0.4222521	total: 1m 15s	remaining: 17m 16s
68:	learn: 0.4214678	total: 1m 16s	remaining: 17m 14s
69:	learn: 0.4207021	total: 1m 17s	remaining: 17m 13s
70:	learn: 0.4197242	total: 1m 18s	remaining: 17m 13s
71:	learn: 0.4182978	total: 1m 20s	remaining: 17m 11s
72:	learn: 0.4165296	total: 1m 21s	remaining: 17m 10s
73:	learn: 0.4158334	total: 1m 22s	remaining: 17m 8s
74:	learn: 0.4143779	total: 1m 23s	remaining: 17m 7s
75:	learn: 0.4131671	total: 1m 24s	remaining: 17m 6s
76:	learn: 0.4122329	total: 1m 25s	remaining: 17m 4s
77:	learn: 0.4113974	total: 1m 26s	remaining: 17m 3s
78:	learn: 0.4103502	total: 1m 27s	r

211:	learn: 0.3352133	total: 3m 54s	remaining: 14m 30s
212:	learn: 0.3348750	total: 3m 55s	remaining: 14m 29s
213:	learn: 0.3345100	total: 3m 56s	remaining: 14m 27s
214:	learn: 0.3340308	total: 3m 57s	remaining: 14m 26s
215:	learn: 0.3336723	total: 3m 58s	remaining: 14m 25s
216:	learn: 0.3332545	total: 3m 59s	remaining: 14m 24s
217:	learn: 0.3328503	total: 4m	remaining: 14m 23s
218:	learn: 0.3323568	total: 4m 1s	remaining: 14m 22s
219:	learn: 0.3319373	total: 4m 2s	remaining: 14m 20s
220:	learn: 0.3316222	total: 4m 3s	remaining: 14m 19s
221:	learn: 0.3311882	total: 4m 4s	remaining: 14m 18s
222:	learn: 0.3308164	total: 4m 6s	remaining: 14m 17s
223:	learn: 0.3303423	total: 4m 7s	remaining: 14m 16s
224:	learn: 0.3300123	total: 4m 8s	remaining: 14m 15s
225:	learn: 0.3295534	total: 4m 9s	remaining: 14m 14s
226:	learn: 0.3286828	total: 4m 10s	remaining: 14m 13s
227:	learn: 0.3282537	total: 4m 11s	remaining: 14m 12s
228:	learn: 0.3276600	total: 4m 12s	remaining: 14m 10s
229:	learn: 0.3273295	

362:	learn: 0.2851111	total: 6m 38s	remaining: 11m 38s
363:	learn: 0.2849055	total: 6m 39s	remaining: 11m 37s
364:	learn: 0.2847049	total: 6m 40s	remaining: 11m 36s
365:	learn: 0.2844855	total: 6m 41s	remaining: 11m 35s
366:	learn: 0.2842918	total: 6m 42s	remaining: 11m 34s
367:	learn: 0.2840362	total: 6m 43s	remaining: 11m 32s
368:	learn: 0.2837931	total: 6m 44s	remaining: 11m 31s
369:	learn: 0.2835880	total: 6m 45s	remaining: 11m 30s
370:	learn: 0.2833948	total: 6m 46s	remaining: 11m 29s
371:	learn: 0.2832304	total: 6m 47s	remaining: 11m 28s
372:	learn: 0.2830001	total: 6m 48s	remaining: 11m 27s
373:	learn: 0.2827466	total: 6m 50s	remaining: 11m 26s
374:	learn: 0.2825141	total: 6m 51s	remaining: 11m 25s
375:	learn: 0.2822708	total: 6m 52s	remaining: 11m 24s
376:	learn: 0.2820393	total: 6m 53s	remaining: 11m 23s
377:	learn: 0.2818766	total: 6m 54s	remaining: 11m 22s
378:	learn: 0.2816604	total: 6m 55s	remaining: 11m 20s
379:	learn: 0.2814403	total: 6m 56s	remaining: 11m 19s
380:	learn

514:	learn: 0.2546993	total: 9m 23s	remaining: 8m 50s
515:	learn: 0.2544850	total: 9m 24s	remaining: 8m 49s
516:	learn: 0.2543133	total: 9m 25s	remaining: 8m 48s
517:	learn: 0.2541190	total: 9m 26s	remaining: 8m 47s
518:	learn: 0.2539647	total: 9m 27s	remaining: 8m 46s
519:	learn: 0.2538351	total: 9m 28s	remaining: 8m 44s
520:	learn: 0.2537361	total: 9m 29s	remaining: 8m 43s
521:	learn: 0.2535652	total: 9m 30s	remaining: 8m 42s
522:	learn: 0.2533873	total: 9m 31s	remaining: 8m 41s
523:	learn: 0.2532308	total: 9m 33s	remaining: 8m 40s
524:	learn: 0.2530595	total: 9m 34s	remaining: 8m 39s
525:	learn: 0.2529637	total: 9m 35s	remaining: 8m 38s
526:	learn: 0.2527915	total: 9m 36s	remaining: 8m 37s
527:	learn: 0.2527121	total: 9m 37s	remaining: 8m 36s
528:	learn: 0.2525946	total: 9m 38s	remaining: 8m 34s
529:	learn: 0.2524571	total: 9m 39s	remaining: 8m 33s
530:	learn: 0.2523094	total: 9m 40s	remaining: 8m 32s
531:	learn: 0.2521483	total: 9m 41s	remaining: 8m 31s
532:	learn: 0.2519960	total:

665:	learn: 0.2331938	total: 12m 9s	remaining: 6m 6s
666:	learn: 0.2330858	total: 12m 11s	remaining: 6m 4s
667:	learn: 0.2329319	total: 12m 12s	remaining: 6m 3s
668:	learn: 0.2328340	total: 12m 13s	remaining: 6m 2s
669:	learn: 0.2327304	total: 12m 14s	remaining: 6m 1s
670:	learn: 0.2325805	total: 12m 15s	remaining: 6m
671:	learn: 0.2325208	total: 12m 16s	remaining: 5m 59s
672:	learn: 0.2323550	total: 12m 17s	remaining: 5m 58s
673:	learn: 0.2322574	total: 12m 19s	remaining: 5m 57s
674:	learn: 0.2321034	total: 12m 20s	remaining: 5m 56s
675:	learn: 0.2319583	total: 12m 21s	remaining: 5m 55s
676:	learn: 0.2318701	total: 12m 22s	remaining: 5m 54s
677:	learn: 0.2315587	total: 12m 23s	remaining: 5m 53s
678:	learn: 0.2314984	total: 12m 24s	remaining: 5m 52s
679:	learn: 0.2313991	total: 12m 25s	remaining: 5m 51s
680:	learn: 0.2313228	total: 12m 27s	remaining: 5m 49s
681:	learn: 0.2311966	total: 12m 28s	remaining: 5m 48s
682:	learn: 0.2310002	total: 12m 29s	remaining: 5m 47s
683:	learn: 0.230886

816:	learn: 0.2159426	total: 15m 5s	remaining: 3m 22s
817:	learn: 0.2158405	total: 15m 7s	remaining: 3m 21s
818:	learn: 0.2157894	total: 15m 8s	remaining: 3m 20s
819:	learn: 0.2156912	total: 15m 9s	remaining: 3m 19s
820:	learn: 0.2156402	total: 15m 10s	remaining: 3m 18s
821:	learn: 0.2155894	total: 15m 11s	remaining: 3m 17s
822:	learn: 0.2154457	total: 15m 12s	remaining: 3m 16s
823:	learn: 0.2153466	total: 15m 13s	remaining: 3m 15s
824:	learn: 0.2151872	total: 15m 14s	remaining: 3m 14s
825:	learn: 0.2150852	total: 15m 15s	remaining: 3m 12s
826:	learn: 0.2150358	total: 15m 16s	remaining: 3m 11s
827:	learn: 0.2149212	total: 15m 17s	remaining: 3m 10s
828:	learn: 0.2148256	total: 15m 18s	remaining: 3m 9s
829:	learn: 0.2147628	total: 15m 19s	remaining: 3m 8s
830:	learn: 0.2147124	total: 15m 20s	remaining: 3m 7s
831:	learn: 0.2146621	total: 15m 22s	remaining: 3m 6s
832:	learn: 0.2145875	total: 15m 23s	remaining: 3m 5s
833:	learn: 0.2145372	total: 15m 25s	remaining: 3m 4s
834:	learn: 0.214407

967:	learn: 0.2024156	total: 18m 11s	remaining: 36.1s
968:	learn: 0.2023058	total: 18m 12s	remaining: 35s
969:	learn: 0.2021473	total: 18m 14s	remaining: 33.8s
970:	learn: 0.2021031	total: 18m 15s	remaining: 32.7s
971:	learn: 0.2019962	total: 18m 16s	remaining: 31.6s
972:	learn: 0.2019163	total: 18m 17s	remaining: 30.4s
973:	learn: 0.2017926	total: 18m 18s	remaining: 29.3s
974:	learn: 0.2017354	total: 18m 19s	remaining: 28.2s
975:	learn: 0.2016325	total: 18m 21s	remaining: 27.1s
976:	learn: 0.2014750	total: 18m 22s	remaining: 26s
977:	learn: 0.2013222	total: 18m 23s	remaining: 24.8s
978:	learn: 0.2012167	total: 18m 25s	remaining: 23.7s
979:	learn: 0.2011265	total: 18m 26s	remaining: 22.6s
980:	learn: 0.2010825	total: 18m 27s	remaining: 21.5s
981:	learn: 0.2009980	total: 18m 29s	remaining: 20.3s
982:	learn: 0.2009555	total: 18m 30s	remaining: 19.2s
983:	learn: 0.2008736	total: 18m 31s	remaining: 18.1s
984:	learn: 0.2007898	total: 18m 32s	remaining: 16.9s
985:	learn: 0.2006989	total: 18m

[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed: 84.4min finished


Learning rate set to 0.079478
0:	learn: 0.6559167	total: 1.66s	remaining: 27m 36s
1:	learn: 0.6270558	total: 3.45s	remaining: 28m 40s
2:	learn: 0.6071170	total: 5.49s	remaining: 30m 23s
3:	learn: 0.5899061	total: 8.12s	remaining: 33m 41s
4:	learn: 0.5767099	total: 10.3s	remaining: 34m 18s
5:	learn: 0.5645835	total: 12s	remaining: 33m 12s
6:	learn: 0.5561485	total: 13.6s	remaining: 32m 16s
7:	learn: 0.5477263	total: 15.2s	remaining: 31m 28s
8:	learn: 0.5411716	total: 16.7s	remaining: 30m 41s
9:	learn: 0.5354903	total: 18.3s	remaining: 30m 9s
10:	learn: 0.5294459	total: 19.8s	remaining: 29m 43s
11:	learn: 0.5221490	total: 21.4s	remaining: 29m 22s
12:	learn: 0.5167986	total: 23.1s	remaining: 29m 10s
13:	learn: 0.5115889	total: 24.6s	remaining: 28m 51s
14:	learn: 0.5076063	total: 26.1s	remaining: 28m 32s
15:	learn: 0.5041646	total: 27.7s	remaining: 28m 20s
16:	learn: 0.4997705	total: 29.1s	remaining: 28m 5s
17:	learn: 0.4966670	total: 30.7s	remaining: 27m 55s
18:	learn: 0.4937544	total: 32

153:	learn: 0.3486688	total: 4m 5s	remaining: 22m 29s
154:	learn: 0.3479995	total: 4m 7s	remaining: 22m 27s
155:	learn: 0.3473135	total: 4m 8s	remaining: 22m 25s
156:	learn: 0.3467873	total: 4m 10s	remaining: 22m 23s
157:	learn: 0.3462556	total: 4m 11s	remaining: 22m 21s
158:	learn: 0.3458133	total: 4m 13s	remaining: 22m 19s
159:	learn: 0.3446864	total: 4m 14s	remaining: 22m 17s
160:	learn: 0.3442501	total: 4m 16s	remaining: 22m 15s
161:	learn: 0.3434199	total: 4m 17s	remaining: 22m 13s
162:	learn: 0.3429757	total: 4m 19s	remaining: 22m 11s
163:	learn: 0.3424307	total: 4m 20s	remaining: 22m 9s
164:	learn: 0.3419518	total: 4m 22s	remaining: 22m 6s
165:	learn: 0.3414764	total: 4m 23s	remaining: 22m 4s
166:	learn: 0.3411360	total: 4m 25s	remaining: 22m 2s
167:	learn: 0.3402969	total: 4m 26s	remaining: 22m 1s
168:	learn: 0.3398779	total: 4m 28s	remaining: 21m 59s
169:	learn: 0.3394214	total: 4m 29s	remaining: 21m 57s
170:	learn: 0.3386087	total: 4m 31s	remaining: 21m 55s
171:	learn: 0.3377

304:	learn: 0.2905281	total: 7m 57s	remaining: 18m 7s
305:	learn: 0.2903231	total: 7m 58s	remaining: 18m 5s
306:	learn: 0.2900173	total: 8m	remaining: 18m 3s
307:	learn: 0.2896345	total: 8m 1s	remaining: 18m 2s
308:	learn: 0.2894078	total: 8m 3s	remaining: 18m
309:	learn: 0.2891823	total: 8m 4s	remaining: 17m 58s
310:	learn: 0.2889247	total: 8m 6s	remaining: 17m 57s
311:	learn: 0.2886850	total: 8m 7s	remaining: 17m 55s
312:	learn: 0.2883998	total: 8m 9s	remaining: 17m 53s
313:	learn: 0.2881651	total: 8m 10s	remaining: 17m 51s
314:	learn: 0.2879479	total: 8m 12s	remaining: 17m 50s
315:	learn: 0.2876437	total: 8m 13s	remaining: 17m 48s
316:	learn: 0.2874473	total: 8m 15s	remaining: 17m 47s
317:	learn: 0.2872767	total: 8m 17s	remaining: 17m 46s
318:	learn: 0.2870234	total: 8m 19s	remaining: 17m 45s
319:	learn: 0.2867400	total: 8m 20s	remaining: 17m 44s
320:	learn: 0.2864780	total: 8m 22s	remaining: 17m 42s
321:	learn: 0.2862650	total: 8m 23s	remaining: 17m 40s
322:	learn: 0.2859971	total:

453:	learn: 0.2581731	total: 11m 58s	remaining: 14m 24s
454:	learn: 0.2579598	total: 12m	remaining: 14m 22s
455:	learn: 0.2578399	total: 12m 1s	remaining: 14m 21s
456:	learn: 0.2576772	total: 12m 3s	remaining: 14m 19s
457:	learn: 0.2574760	total: 12m 5s	remaining: 14m 18s
458:	learn: 0.2572632	total: 12m 6s	remaining: 14m 16s
459:	learn: 0.2571229	total: 12m 8s	remaining: 14m 14s
460:	learn: 0.2569638	total: 12m 9s	remaining: 14m 13s
461:	learn: 0.2568080	total: 12m 11s	remaining: 14m 11s
462:	learn: 0.2566477	total: 12m 13s	remaining: 14m 10s
463:	learn: 0.2564272	total: 12m 15s	remaining: 14m 9s
464:	learn: 0.2562442	total: 12m 17s	remaining: 14m 8s
465:	learn: 0.2560780	total: 12m 18s	remaining: 14m 6s
466:	learn: 0.2559019	total: 12m 20s	remaining: 14m 5s
467:	learn: 0.2557700	total: 12m 22s	remaining: 14m 3s
468:	learn: 0.2555786	total: 12m 24s	remaining: 14m 2s
469:	learn: 0.2554220	total: 12m 25s	remaining: 14m
470:	learn: 0.2553217	total: 12m 27s	remaining: 13m 59s
471:	learn: 

601:	learn: 0.2364913	total: 15m 48s	remaining: 10m 27s
602:	learn: 0.2363783	total: 15m 50s	remaining: 10m 25s
603:	learn: 0.2362447	total: 15m 52s	remaining: 10m 24s
604:	learn: 0.2361721	total: 15m 53s	remaining: 10m 22s
605:	learn: 0.2360553	total: 15m 55s	remaining: 10m 21s
606:	learn: 0.2359946	total: 15m 56s	remaining: 10m 19s
607:	learn: 0.2358964	total: 15m 58s	remaining: 10m 17s
608:	learn: 0.2357998	total: 15m 59s	remaining: 10m 16s
609:	learn: 0.2356757	total: 16m 1s	remaining: 10m 14s
610:	learn: 0.2355241	total: 16m 2s	remaining: 10m 13s
611:	learn: 0.2354153	total: 16m 4s	remaining: 10m 11s
612:	learn: 0.2352916	total: 16m 6s	remaining: 10m 9s
613:	learn: 0.2351151	total: 16m 7s	remaining: 10m 8s
614:	learn: 0.2350195	total: 16m 8s	remaining: 10m 6s
615:	learn: 0.2348905	total: 16m 10s	remaining: 10m 4s
616:	learn: 0.2347677	total: 16m 11s	remaining: 10m 3s
617:	learn: 0.2346949	total: 16m 13s	remaining: 10m 1s
618:	learn: 0.2345756	total: 16m 14s	remaining: 9m 59s
619:	

751:	learn: 0.2191956	total: 19m 41s	remaining: 6m 29s
752:	learn: 0.2190920	total: 19m 43s	remaining: 6m 28s
753:	learn: 0.2190430	total: 19m 44s	remaining: 6m 26s
754:	learn: 0.2189785	total: 19m 46s	remaining: 6m 25s
755:	learn: 0.2189310	total: 19m 48s	remaining: 6m 23s
756:	learn: 0.2188307	total: 19m 49s	remaining: 6m 21s
757:	learn: 0.2187254	total: 19m 51s	remaining: 6m 20s
758:	learn: 0.2186057	total: 19m 52s	remaining: 6m 18s
759:	learn: 0.2184981	total: 19m 54s	remaining: 6m 17s
760:	learn: 0.2183934	total: 19m 56s	remaining: 6m 15s
761:	learn: 0.2182921	total: 19m 57s	remaining: 6m 14s
762:	learn: 0.2182062	total: 19m 59s	remaining: 6m 12s
763:	learn: 0.2181050	total: 20m 1s	remaining: 6m 11s
764:	learn: 0.2180286	total: 20m 2s	remaining: 6m 9s
765:	learn: 0.2179288	total: 20m 4s	remaining: 6m 7s
766:	learn: 0.2178807	total: 20m 6s	remaining: 6m 6s
767:	learn: 0.2177654	total: 20m 7s	remaining: 6m 4s
768:	learn: 0.2176659	total: 20m 9s	remaining: 6m 3s
769:	learn: 0.2175716

902:	learn: 0.2056781	total: 23m 36s	remaining: 2m 32s
903:	learn: 0.2055860	total: 23m 38s	remaining: 2m 30s
904:	learn: 0.2055461	total: 23m 40s	remaining: 2m 29s
905:	learn: 0.2054616	total: 23m 41s	remaining: 2m 27s
906:	learn: 0.2053744	total: 23m 43s	remaining: 2m 25s
907:	learn: 0.2052829	total: 23m 45s	remaining: 2m 24s
908:	learn: 0.2051943	total: 23m 46s	remaining: 2m 22s
909:	learn: 0.2051529	total: 23m 48s	remaining: 2m 21s
910:	learn: 0.2050314	total: 23m 50s	remaining: 2m 19s
911:	learn: 0.2049767	total: 23m 52s	remaining: 2m 18s
912:	learn: 0.2048457	total: 23m 54s	remaining: 2m 16s
913:	learn: 0.2048044	total: 23m 56s	remaining: 2m 15s
914:	learn: 0.2047631	total: 23m 58s	remaining: 2m 13s
915:	learn: 0.2047219	total: 23m 59s	remaining: 2m 12s
916:	learn: 0.2046393	total: 24m 1s	remaining: 2m 10s
917:	learn: 0.2045525	total: 24m 2s	remaining: 2m 8s
918:	learn: 0.2044770	total: 24m 4s	remaining: 2m 7s
919:	learn: 0.2044359	total: 24m 5s	remaining: 2m 5s
920:	learn: 0.204

GridSearchCV(cv=3, error_score='raise-deprecating',
             estimator=<catboost.core.CatBoostClassifier object at 0x000000002CC075C8>,
             iid='warn', n_jobs=None, param_grid={'iterations': [500, 1000]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring='f1', verbose=5)

In [29]:
print(f'Лучшие параметры случайного леса - {grid_cat.best_params_}, score = {grid_cat.best_score_}')

Лучшие параметры случайного леса - {'iterations': 1000}, score = 0.750951538815851


F1 для обоих моделей больше 0,75, проверим качество предсказаний на тестовой выборке:

In [30]:
def test_validate(model, name):
    print('*' * 45)
    print(f'Проверка качества модели {name} на тестовой выборке')
    model.fit(X_train_vec, y_train)
    predict = model.predict(X_test_vec)
    print(f'f1_score предсказаний => {f1_score(y_test, predict)}')
    print()
    print('*' * 45)

In [31]:
for i in ((grid_log.best_estimator_, 'Логистическая регрессия'), (grid_cat.best_estimator_, 'CatBoost')):
    test_validate(i[0], i[1])

*********************************************
Проверка качества модели Логистическая регрессия на тестовой выборке
f1_score предсказаний => 0.7680860805860805

*********************************************
*********************************************
Проверка качества модели CatBoost на тестовой выборке
Learning rate set to 0.079478
0:	learn: 0.6559167	total: 1.57s	remaining: 26m 14s
1:	learn: 0.6270558	total: 3.09s	remaining: 25m 41s
2:	learn: 0.6071170	total: 4.62s	remaining: 25m 33s
3:	learn: 0.5899061	total: 6.17s	remaining: 25m 37s
4:	learn: 0.5767099	total: 7.71s	remaining: 25m 35s
5:	learn: 0.5645835	total: 9.28s	remaining: 25m 37s
6:	learn: 0.5561485	total: 10.9s	remaining: 25m 49s
7:	learn: 0.5477263	total: 12.5s	remaining: 25m 48s
8:	learn: 0.5411716	total: 14s	remaining: 25m 42s
9:	learn: 0.5354903	total: 15.6s	remaining: 25m 43s
10:	learn: 0.5294459	total: 17.1s	remaining: 25m 39s
11:	learn: 0.5221490	total: 18.7s	remaining: 25m 36s
12:	learn: 0.5167986	total: 20.2s	remain

147:	learn: 0.3516518	total: 3m 52s	remaining: 22m 18s
148:	learn: 0.3511534	total: 3m 54s	remaining: 22m 17s
149:	learn: 0.3506777	total: 3m 55s	remaining: 22m 16s
150:	learn: 0.3501511	total: 3m 57s	remaining: 22m 14s
151:	learn: 0.3497117	total: 3m 58s	remaining: 22m 12s
152:	learn: 0.3492222	total: 4m	remaining: 22m 11s
153:	learn: 0.3486688	total: 4m 2s	remaining: 22m 9s
154:	learn: 0.3479995	total: 4m 3s	remaining: 22m 8s
155:	learn: 0.3473135	total: 4m 5s	remaining: 22m 7s
156:	learn: 0.3467873	total: 4m 6s	remaining: 22m 5s
157:	learn: 0.3462556	total: 4m 8s	remaining: 22m 4s
158:	learn: 0.3458133	total: 4m 10s	remaining: 22m 3s
159:	learn: 0.3446864	total: 4m 11s	remaining: 22m 2s
160:	learn: 0.3442501	total: 4m 13s	remaining: 22m
161:	learn: 0.3434199	total: 4m 14s	remaining: 21m 59s
162:	learn: 0.3429757	total: 4m 16s	remaining: 21m 57s
163:	learn: 0.3424307	total: 4m 18s	remaining: 21m 56s
164:	learn: 0.3419518	total: 4m 19s	remaining: 21m 54s
165:	learn: 0.3414764	total: 4

298:	learn: 0.2919931	total: 7m 54s	remaining: 18m 32s
299:	learn: 0.2917574	total: 7m 55s	remaining: 18m 30s
300:	learn: 0.2915457	total: 7m 57s	remaining: 18m 28s
301:	learn: 0.2912946	total: 7m 59s	remaining: 18m 27s
302:	learn: 0.2910631	total: 8m	remaining: 18m 25s
303:	learn: 0.2907980	total: 8m 2s	remaining: 18m 23s
304:	learn: 0.2905281	total: 8m 3s	remaining: 18m 22s
305:	learn: 0.2903231	total: 8m 5s	remaining: 18m 20s
306:	learn: 0.2900173	total: 8m 7s	remaining: 18m 19s
307:	learn: 0.2896345	total: 8m 8s	remaining: 18m 18s
308:	learn: 0.2894078	total: 8m 10s	remaining: 18m 16s
309:	learn: 0.2891823	total: 8m 11s	remaining: 18m 14s
310:	learn: 0.2889247	total: 8m 13s	remaining: 18m 13s
311:	learn: 0.2886850	total: 8m 14s	remaining: 18m 11s
312:	learn: 0.2883998	total: 8m 16s	remaining: 18m 9s
313:	learn: 0.2881651	total: 8m 18s	remaining: 18m 8s
314:	learn: 0.2879479	total: 8m 19s	remaining: 18m 6s
315:	learn: 0.2876437	total: 8m 21s	remaining: 18m 4s
316:	learn: 0.2874473	t

447:	learn: 0.2597455	total: 11m 54s	remaining: 14m 40s
448:	learn: 0.2595905	total: 11m 56s	remaining: 14m 38s
449:	learn: 0.2594471	total: 11m 57s	remaining: 14m 37s
450:	learn: 0.2593480	total: 11m 59s	remaining: 14m 35s
451:	learn: 0.2592025	total: 12m	remaining: 14m 33s
452:	learn: 0.2588644	total: 12m 2s	remaining: 14m 32s
453:	learn: 0.2581731	total: 12m 3s	remaining: 14m 30s
454:	learn: 0.2579598	total: 12m 5s	remaining: 14m 29s
455:	learn: 0.2578399	total: 12m 7s	remaining: 14m 27s
456:	learn: 0.2576772	total: 12m 8s	remaining: 14m 25s
457:	learn: 0.2574760	total: 12m 10s	remaining: 14m 24s
458:	learn: 0.2572632	total: 12m 11s	remaining: 14m 22s
459:	learn: 0.2571229	total: 12m 13s	remaining: 14m 20s
460:	learn: 0.2569638	total: 12m 15s	remaining: 14m 19s
461:	learn: 0.2568080	total: 12m 16s	remaining: 14m 17s
462:	learn: 0.2566477	total: 12m 18s	remaining: 14m 16s
463:	learn: 0.2564272	total: 12m 19s	remaining: 14m 14s
464:	learn: 0.2562442	total: 12m 21s	remaining: 14m 12s
4

595:	learn: 0.2373585	total: 15m 38s	remaining: 10m 36s
596:	learn: 0.2372555	total: 15m 39s	remaining: 10m 34s
597:	learn: 0.2370404	total: 15m 41s	remaining: 10m 33s
598:	learn: 0.2368592	total: 15m 43s	remaining: 10m 31s
599:	learn: 0.2367058	total: 15m 45s	remaining: 10m 30s
600:	learn: 0.2365764	total: 15m 47s	remaining: 10m 29s
601:	learn: 0.2364913	total: 15m 49s	remaining: 10m 27s
602:	learn: 0.2363783	total: 15m 51s	remaining: 10m 26s
603:	learn: 0.2362447	total: 15m 53s	remaining: 10m 24s
604:	learn: 0.2361721	total: 15m 54s	remaining: 10m 23s
605:	learn: 0.2360553	total: 15m 56s	remaining: 10m 21s
606:	learn: 0.2359946	total: 15m 58s	remaining: 10m 20s
607:	learn: 0.2358964	total: 15m 59s	remaining: 10m 18s
608:	learn: 0.2357998	total: 16m 1s	remaining: 10m 17s
609:	learn: 0.2356757	total: 16m 2s	remaining: 10m 15s
610:	learn: 0.2355241	total: 16m 4s	remaining: 10m 13s
611:	learn: 0.2354153	total: 16m 5s	remaining: 10m 12s
612:	learn: 0.2352916	total: 16m 7s	remaining: 10m 1

745:	learn: 0.2197608	total: 19m 39s	remaining: 6m 41s
746:	learn: 0.2196829	total: 19m 40s	remaining: 6m 39s
747:	learn: 0.2195681	total: 19m 42s	remaining: 6m 38s
748:	learn: 0.2195010	total: 19m 43s	remaining: 6m 36s
749:	learn: 0.2194281	total: 19m 45s	remaining: 6m 35s
750:	learn: 0.2192921	total: 19m 46s	remaining: 6m 33s
751:	learn: 0.2191956	total: 19m 48s	remaining: 6m 31s
752:	learn: 0.2190920	total: 19m 49s	remaining: 6m 30s
753:	learn: 0.2190430	total: 19m 51s	remaining: 6m 28s
754:	learn: 0.2189785	total: 19m 52s	remaining: 6m 26s
755:	learn: 0.2189310	total: 19m 54s	remaining: 6m 25s
756:	learn: 0.2188307	total: 19m 55s	remaining: 6m 23s
757:	learn: 0.2187254	total: 19m 56s	remaining: 6m 22s
758:	learn: 0.2186057	total: 19m 58s	remaining: 6m 20s
759:	learn: 0.2184981	total: 19m 59s	remaining: 6m 18s
760:	learn: 0.2183934	total: 20m 1s	remaining: 6m 17s
761:	learn: 0.2182921	total: 20m 2s	remaining: 6m 15s
762:	learn: 0.2182062	total: 20m 4s	remaining: 6m 14s
763:	learn: 0

896:	learn: 0.2061044	total: 23m 32s	remaining: 2m 42s
897:	learn: 0.2060133	total: 23m 33s	remaining: 2m 40s
898:	learn: 0.2059719	total: 23m 35s	remaining: 2m 39s
899:	learn: 0.2058756	total: 23m 36s	remaining: 2m 37s
900:	learn: 0.2058024	total: 23m 38s	remaining: 2m 35s
901:	learn: 0.2057290	total: 23m 39s	remaining: 2m 34s
902:	learn: 0.2056781	total: 23m 41s	remaining: 2m 32s
903:	learn: 0.2055860	total: 23m 42s	remaining: 2m 31s
904:	learn: 0.2055461	total: 23m 44s	remaining: 2m 29s
905:	learn: 0.2054616	total: 23m 45s	remaining: 2m 27s
906:	learn: 0.2053744	total: 23m 47s	remaining: 2m 26s
907:	learn: 0.2052829	total: 23m 48s	remaining: 2m 24s
908:	learn: 0.2051943	total: 23m 50s	remaining: 2m 23s
909:	learn: 0.2051529	total: 23m 51s	remaining: 2m 21s
910:	learn: 0.2050314	total: 23m 53s	remaining: 2m 20s
911:	learn: 0.2049767	total: 23m 54s	remaining: 2m 18s
912:	learn: 0.2048457	total: 23m 56s	remaining: 2m 16s
913:	learn: 0.2048044	total: 23m 57s	remaining: 2m 15s
914:	learn

### Выводы:

1. Рассмотрены модели логистической регрессии и бустер Catboost, модели обучены и получены удовлетворяющие условию проекта (f1 > 0.75) как на кроссвалидации, так и на тестовой выборке.
2. Логистическая регрессия на 0.8% лучше модели Catboost по метрике f1. Разница на мой взгляд небольшая, но, так как скорость обучения модели логистической регрессии гораздо выше, я бы выбрал эту модель.

## 2.4 Обучение моделей на полном наборе данных и сравнение значений F1 score

Объединим обучающие и тестовые признаки и целевой признак:

In [74]:
X = vstack([X_train_vec, X_test_vec])
y = pd.concat([y_train, y_test])

Без этого объединения будет не правильно делать выводы о подобраных гиперпараметрах, так как размерность признаков изменится

Обучим модели и посмотрим на значение метрики f1

**Логистическая регрессия**

In [77]:
best_mod_log = grid_log.best_estimator_

In [79]:
def f1_calc(model):
    print(f'Значение метрики f1 => {cross_val_score(model, X, y, cv=3, scoring="f1").mean()}')

In [85]:
%%time
f1_calc(best_mod_log)

Значение метрики f1 => 0.7618427627189299
Wall time: 8.7 s


**Модель Catboost**

In [81]:
best_mod_cat = grid_cat.best_estimator_

In [82]:
f1_calc(best_mod_cat)

Learning rate set to 0.075579
0:	learn: 0.6561934	total: 1.44s	remaining: 23m 55s
1:	learn: 0.6293006	total: 2.87s	remaining: 23m 50s
2:	learn: 0.6072333	total: 4.17s	remaining: 23m 5s
3:	learn: 0.5901452	total: 5.52s	remaining: 22m 54s
4:	learn: 0.5763682	total: 6.83s	remaining: 22m 38s
5:	learn: 0.5657718	total: 8.11s	remaining: 22m 23s
6:	learn: 0.5568693	total: 9.43s	remaining: 22m 17s
7:	learn: 0.5509795	total: 10.7s	remaining: 22m 12s
8:	learn: 0.5435811	total: 12s	remaining: 22m 6s
9:	learn: 0.5375687	total: 13.4s	remaining: 22m 9s
10:	learn: 0.5313564	total: 14.7s	remaining: 22m 2s
11:	learn: 0.5234747	total: 16s	remaining: 21m 58s
12:	learn: 0.5181725	total: 17.3s	remaining: 21m 54s
13:	learn: 0.5144778	total: 18.6s	remaining: 21m 51s
14:	learn: 0.5100040	total: 19.9s	remaining: 21m 49s
15:	learn: 0.5063130	total: 21.3s	remaining: 21m 49s
16:	learn: 0.5026765	total: 22.6s	remaining: 21m 45s
17:	learn: 0.4987745	total: 23.9s	remaining: 21m 41s
18:	learn: 0.4950614	total: 25.2s	

153:	learn: 0.3479922	total: 3m 17s	remaining: 18m 7s
154:	learn: 0.3475065	total: 3m 19s	remaining: 18m 6s
155:	learn: 0.3469380	total: 3m 20s	remaining: 18m 4s
156:	learn: 0.3465135	total: 3m 21s	remaining: 18m 3s
157:	learn: 0.3461090	total: 3m 22s	remaining: 18m 1s
158:	learn: 0.3456383	total: 3m 24s	remaining: 18m
159:	learn: 0.3452068	total: 3m 25s	remaining: 17m 58s
160:	learn: 0.3447149	total: 3m 26s	remaining: 17m 57s
161:	learn: 0.3441968	total: 3m 28s	remaining: 17m 55s
162:	learn: 0.3437640	total: 3m 29s	remaining: 17m 54s
163:	learn: 0.3431210	total: 3m 30s	remaining: 17m 53s
164:	learn: 0.3426367	total: 3m 31s	remaining: 17m 51s
165:	learn: 0.3422944	total: 3m 33s	remaining: 17m 50s
166:	learn: 0.3418351	total: 3m 34s	remaining: 17m 48s
167:	learn: 0.3410051	total: 3m 35s	remaining: 17m 47s
168:	learn: 0.3405302	total: 3m 36s	remaining: 17m 46s
169:	learn: 0.3391178	total: 3m 38s	remaining: 17m 44s
170:	learn: 0.3387752	total: 3m 39s	remaining: 17m 43s
171:	learn: 0.33824

304:	learn: 0.2889170	total: 6m 40s	remaining: 15m 11s
305:	learn: 0.2886576	total: 6m 41s	remaining: 15m 10s
306:	learn: 0.2884693	total: 6m 42s	remaining: 15m 9s
307:	learn: 0.2882330	total: 6m 44s	remaining: 15m 7s
308:	learn: 0.2879906	total: 6m 45s	remaining: 15m 6s
309:	learn: 0.2877295	total: 6m 47s	remaining: 15m 6s
310:	learn: 0.2875278	total: 6m 48s	remaining: 15m 4s
311:	learn: 0.2872606	total: 6m 49s	remaining: 15m 3s
312:	learn: 0.2869931	total: 6m 51s	remaining: 15m 2s
313:	learn: 0.2866784	total: 6m 52s	remaining: 15m 1s
314:	learn: 0.2864974	total: 6m 53s	remaining: 14m 59s
315:	learn: 0.2862903	total: 6m 55s	remaining: 14m 58s
316:	learn: 0.2860435	total: 6m 56s	remaining: 14m 57s
317:	learn: 0.2858398	total: 6m 57s	remaining: 14m 55s
318:	learn: 0.2855131	total: 6m 59s	remaining: 14m 54s
319:	learn: 0.2852845	total: 7m	remaining: 14m 53s
320:	learn: 0.2850679	total: 7m 1s	remaining: 14m 51s
321:	learn: 0.2848558	total: 7m 2s	remaining: 14m 50s
322:	learn: 0.2846325	to

455:	learn: 0.2566063	total: 10m 1s	remaining: 11m 57s
456:	learn: 0.2563867	total: 10m 2s	remaining: 11m 55s
457:	learn: 0.2562311	total: 10m 3s	remaining: 11m 54s
458:	learn: 0.2560865	total: 10m 4s	remaining: 11m 53s
459:	learn: 0.2559794	total: 10m 6s	remaining: 11m 51s
460:	learn: 0.2558382	total: 10m 7s	remaining: 11m 50s
461:	learn: 0.2556588	total: 10m 8s	remaining: 11m 49s
462:	learn: 0.2554695	total: 10m 10s	remaining: 11m 47s
463:	learn: 0.2553147	total: 10m 11s	remaining: 11m 46s
464:	learn: 0.2551848	total: 10m 12s	remaining: 11m 44s
465:	learn: 0.2549805	total: 10m 13s	remaining: 11m 43s
466:	learn: 0.2547779	total: 10m 15s	remaining: 11m 42s
467:	learn: 0.2545973	total: 10m 16s	remaining: 11m 40s
468:	learn: 0.2544593	total: 10m 17s	remaining: 11m 39s
469:	learn: 0.2542234	total: 10m 19s	remaining: 11m 38s
470:	learn: 0.2540222	total: 10m 20s	remaining: 11m 36s
471:	learn: 0.2537838	total: 10m 21s	remaining: 11m 35s
472:	learn: 0.2534150	total: 10m 22s	remaining: 11m 33s

604:	learn: 0.2338154	total: 13m 14s	remaining: 8m 38s
605:	learn: 0.2337393	total: 13m 16s	remaining: 8m 37s
606:	learn: 0.2334367	total: 13m 17s	remaining: 8m 36s
607:	learn: 0.2332671	total: 13m 19s	remaining: 8m 35s
608:	learn: 0.2332070	total: 13m 21s	remaining: 8m 34s
609:	learn: 0.2330495	total: 13m 22s	remaining: 8m 33s
610:	learn: 0.2329455	total: 13m 24s	remaining: 8m 31s
611:	learn: 0.2328779	total: 13m 25s	remaining: 8m 30s
612:	learn: 0.2327271	total: 13m 26s	remaining: 8m 29s
613:	learn: 0.2326579	total: 13m 28s	remaining: 8m 28s
614:	learn: 0.2325054	total: 13m 29s	remaining: 8m 26s
615:	learn: 0.2323740	total: 13m 30s	remaining: 8m 25s
616:	learn: 0.2321846	total: 13m 31s	remaining: 8m 24s
617:	learn: 0.2320392	total: 13m 33s	remaining: 8m 22s
618:	learn: 0.2318962	total: 13m 34s	remaining: 8m 21s
619:	learn: 0.2318124	total: 13m 35s	remaining: 8m 20s
620:	learn: 0.2316765	total: 13m 37s	remaining: 8m 18s
621:	learn: 0.2315333	total: 13m 38s	remaining: 8m 17s
622:	learn

755:	learn: 0.2163755	total: 16m 34s	remaining: 5m 21s
756:	learn: 0.2162872	total: 16m 35s	remaining: 5m 19s
757:	learn: 0.2161892	total: 16m 37s	remaining: 5m 18s
758:	learn: 0.2159949	total: 16m 38s	remaining: 5m 17s
759:	learn: 0.2158762	total: 16m 40s	remaining: 5m 15s
760:	learn: 0.2157872	total: 16m 41s	remaining: 5m 14s
761:	learn: 0.2157037	total: 16m 43s	remaining: 5m 13s
762:	learn: 0.2156558	total: 16m 44s	remaining: 5m 11s
763:	learn: 0.2155676	total: 16m 45s	remaining: 5m 10s
764:	learn: 0.2154392	total: 16m 46s	remaining: 5m 9s
765:	learn: 0.2153614	total: 16m 48s	remaining: 5m 7s
766:	learn: 0.2152299	total: 16m 49s	remaining: 5m 6s
767:	learn: 0.2151165	total: 16m 50s	remaining: 5m 5s
768:	learn: 0.2150574	total: 16m 51s	remaining: 5m 3s
769:	learn: 0.2149581	total: 16m 53s	remaining: 5m 2s
770:	learn: 0.2148417	total: 16m 54s	remaining: 5m 1s
771:	learn: 0.2147222	total: 16m 55s	remaining: 4m 59s
772:	learn: 0.2146054	total: 16m 57s	remaining: 4m 58s
773:	learn: 0.214

906:	learn: 0.2029221	total: 19m 45s	remaining: 2m 1s
907:	learn: 0.2028304	total: 19m 47s	remaining: 2m
908:	learn: 0.2027350	total: 19m 48s	remaining: 1m 58s
909:	learn: 0.2026533	total: 19m 49s	remaining: 1m 57s
910:	learn: 0.2026107	total: 19m 50s	remaining: 1m 56s
911:	learn: 0.2025567	total: 19m 52s	remaining: 1m 55s
912:	learn: 0.2025145	total: 19m 53s	remaining: 1m 53s
913:	learn: 0.2024338	total: 19m 54s	remaining: 1m 52s
914:	learn: 0.2022510	total: 19m 55s	remaining: 1m 51s
915:	learn: 0.2022106	total: 19m 57s	remaining: 1m 49s
916:	learn: 0.2021697	total: 19m 58s	remaining: 1m 48s
917:	learn: 0.2020912	total: 19m 59s	remaining: 1m 47s
918:	learn: 0.2020498	total: 20m	remaining: 1m 45s
919:	learn: 0.2019706	total: 20m 2s	remaining: 1m 44s
920:	learn: 0.2019295	total: 20m 3s	remaining: 1m 43s
921:	learn: 0.2018232	total: 20m 4s	remaining: 1m 41s
922:	learn: 0.2017823	total: 20m 5s	remaining: 1m 40s
923:	learn: 0.2016841	total: 20m 7s	remaining: 1m 39s
924:	learn: 0.2015868	to

59:	learn: 0.4177430	total: 1m 19s	remaining: 20m 42s
60:	learn: 0.4164737	total: 1m 20s	remaining: 20m 40s
61:	learn: 0.4154857	total: 1m 22s	remaining: 20m 40s
62:	learn: 0.4143660	total: 1m 23s	remaining: 20m 40s
63:	learn: 0.4132645	total: 1m 24s	remaining: 20m 39s
64:	learn: 0.4116281	total: 1m 26s	remaining: 20m 38s
65:	learn: 0.4108808	total: 1m 27s	remaining: 20m 38s
66:	learn: 0.4098031	total: 1m 28s	remaining: 20m 38s
67:	learn: 0.4087777	total: 1m 30s	remaining: 20m 36s
68:	learn: 0.4080609	total: 1m 31s	remaining: 20m 35s
69:	learn: 0.4070380	total: 1m 32s	remaining: 20m 33s
70:	learn: 0.4062351	total: 1m 34s	remaining: 20m 35s
71:	learn: 0.4056047	total: 1m 36s	remaining: 20m 42s
72:	learn: 0.4050028	total: 1m 38s	remaining: 20m 44s
73:	learn: 0.4035145	total: 1m 39s	remaining: 20m 42s
74:	learn: 0.4027913	total: 1m 40s	remaining: 20m 40s
75:	learn: 0.4018499	total: 1m 41s	remaining: 20m 38s
76:	learn: 0.4010635	total: 1m 43s	remaining: 20m 39s
77:	learn: 0.4004429	total: 

210:	learn: 0.3256948	total: 4m 35s	remaining: 17m 8s
211:	learn: 0.3252164	total: 4m 36s	remaining: 17m 7s
212:	learn: 0.3247900	total: 4m 37s	remaining: 17m 5s
213:	learn: 0.3244153	total: 4m 38s	remaining: 17m 4s
214:	learn: 0.3240775	total: 4m 40s	remaining: 17m 3s
215:	learn: 0.3236777	total: 4m 41s	remaining: 17m 1s
216:	learn: 0.3233505	total: 4m 42s	remaining: 17m
217:	learn: 0.3230515	total: 4m 44s	remaining: 16m 59s
218:	learn: 0.3227441	total: 4m 45s	remaining: 16m 57s
219:	learn: 0.3219289	total: 4m 46s	remaining: 16m 56s
220:	learn: 0.3209098	total: 4m 47s	remaining: 16m 54s
221:	learn: 0.3205676	total: 4m 49s	remaining: 16m 53s
222:	learn: 0.3201062	total: 4m 50s	remaining: 16m 52s
223:	learn: 0.3197717	total: 4m 51s	remaining: 16m 50s
224:	learn: 0.3194296	total: 4m 53s	remaining: 16m 49s
225:	learn: 0.3190438	total: 4m 54s	remaining: 16m 47s
226:	learn: 0.3186491	total: 4m 55s	remaining: 16m 46s
227:	learn: 0.3183212	total: 4m 56s	remaining: 16m 45s
228:	learn: 0.317843

361:	learn: 0.2793056	total: 7m 49s	remaining: 13m 47s
362:	learn: 0.2790619	total: 7m 50s	remaining: 13m 46s
363:	learn: 0.2788254	total: 7m 52s	remaining: 13m 44s
364:	learn: 0.2786422	total: 7m 53s	remaining: 13m 43s
365:	learn: 0.2784657	total: 7m 54s	remaining: 13m 41s
366:	learn: 0.2782755	total: 7m 55s	remaining: 13m 40s
367:	learn: 0.2780319	total: 7m 57s	remaining: 13m 39s
368:	learn: 0.2777681	total: 7m 58s	remaining: 13m 37s
369:	learn: 0.2775580	total: 7m 59s	remaining: 13m 36s
370:	learn: 0.2773586	total: 8m	remaining: 13m 35s
371:	learn: 0.2771078	total: 8m 2s	remaining: 13m 34s
372:	learn: 0.2768736	total: 8m 3s	remaining: 13m 32s
373:	learn: 0.2767385	total: 8m 4s	remaining: 13m 31s
374:	learn: 0.2762842	total: 8m 6s	remaining: 13m 30s
375:	learn: 0.2759583	total: 8m 7s	remaining: 13m 28s
376:	learn: 0.2757337	total: 8m 8s	remaining: 13m 27s
377:	learn: 0.2755453	total: 8m 9s	remaining: 13m 26s
378:	learn: 0.2753281	total: 8m 11s	remaining: 13m 24s
379:	learn: 0.2751591

511:	learn: 0.2505705	total: 11m	remaining: 10m 29s
512:	learn: 0.2504385	total: 11m 1s	remaining: 10m 28s
513:	learn: 0.2502849	total: 11m 3s	remaining: 10m 27s
514:	learn: 0.2501443	total: 11m 4s	remaining: 10m 25s
515:	learn: 0.2499758	total: 11m 5s	remaining: 10m 24s
516:	learn: 0.2498275	total: 11m 7s	remaining: 10m 23s
517:	learn: 0.2497114	total: 11m 8s	remaining: 10m 22s
518:	learn: 0.2495207	total: 11m 9s	remaining: 10m 20s
519:	learn: 0.2493766	total: 11m 11s	remaining: 10m 19s
520:	learn: 0.2492706	total: 11m 12s	remaining: 10m 18s
521:	learn: 0.2490771	total: 11m 13s	remaining: 10m 16s
522:	learn: 0.2489395	total: 11m 14s	remaining: 10m 15s
523:	learn: 0.2487879	total: 11m 16s	remaining: 10m 14s
524:	learn: 0.2486974	total: 11m 17s	remaining: 10m 12s
525:	learn: 0.2485422	total: 11m 18s	remaining: 10m 11s
526:	learn: 0.2483884	total: 11m 19s	remaining: 10m 10s
527:	learn: 0.2482383	total: 11m 21s	remaining: 10m 8s
528:	learn: 0.2480313	total: 11m 22s	remaining: 10m 7s
529:	

661:	learn: 0.2302897	total: 14m 10s	remaining: 7m 14s
662:	learn: 0.2301633	total: 14m 12s	remaining: 7m 13s
663:	learn: 0.2300161	total: 14m 13s	remaining: 7m 11s
664:	learn: 0.2299601	total: 14m 14s	remaining: 7m 10s
665:	learn: 0.2298702	total: 14m 15s	remaining: 7m 9s
666:	learn: 0.2297047	total: 14m 17s	remaining: 7m 7s
667:	learn: 0.2295922	total: 14m 18s	remaining: 7m 6s
668:	learn: 0.2294568	total: 14m 19s	remaining: 7m 5s
669:	learn: 0.2293803	total: 14m 21s	remaining: 7m 4s
670:	learn: 0.2292714	total: 14m 22s	remaining: 7m 2s
671:	learn: 0.2292170	total: 14m 23s	remaining: 7m 1s
672:	learn: 0.2290664	total: 14m 24s	remaining: 7m
673:	learn: 0.2289180	total: 14m 25s	remaining: 6m 58s
674:	learn: 0.2287765	total: 14m 27s	remaining: 6m 57s
675:	learn: 0.2286568	total: 14m 28s	remaining: 6m 56s
676:	learn: 0.2285252	total: 14m 29s	remaining: 6m 54s
677:	learn: 0.2284003	total: 14m 31s	remaining: 6m 53s
678:	learn: 0.2282583	total: 14m 32s	remaining: 6m 52s
679:	learn: 0.2280903

812:	learn: 0.2140594	total: 17m 26s	remaining: 4m
813:	learn: 0.2139896	total: 17m 28s	remaining: 3m 59s
814:	learn: 0.2138793	total: 17m 29s	remaining: 3m 58s
815:	learn: 0.2138026	total: 17m 30s	remaining: 3m 56s
816:	learn: 0.2136731	total: 17m 32s	remaining: 3m 55s
817:	learn: 0.2135958	total: 17m 33s	remaining: 3m 54s
818:	learn: 0.2135496	total: 17m 34s	remaining: 3m 53s
819:	learn: 0.2134332	total: 17m 36s	remaining: 3m 51s
820:	learn: 0.2133321	total: 17m 37s	remaining: 3m 50s
821:	learn: 0.2132139	total: 17m 38s	remaining: 3m 49s
822:	learn: 0.2131677	total: 17m 39s	remaining: 3m 47s
823:	learn: 0.2130687	total: 17m 41s	remaining: 3m 46s
824:	learn: 0.2130114	total: 17m 42s	remaining: 3m 45s
825:	learn: 0.2129166	total: 17m 43s	remaining: 3m 44s
826:	learn: 0.2128439	total: 17m 45s	remaining: 3m 42s
827:	learn: 0.2127823	total: 17m 46s	remaining: 3m 41s
828:	learn: 0.2126782	total: 17m 47s	remaining: 3m 40s
829:	learn: 0.2126333	total: 17m 49s	remaining: 3m 39s
830:	learn: 0.

963:	learn: 0.2009257	total: 20m 42s	remaining: 46.4s
964:	learn: 0.2008211	total: 20m 43s	remaining: 45.1s
965:	learn: 0.2007295	total: 20m 44s	remaining: 43.8s
966:	learn: 0.2006792	total: 20m 46s	remaining: 42.5s
967:	learn: 0.2005568	total: 20m 47s	remaining: 41.2s
968:	learn: 0.2004880	total: 20m 48s	remaining: 40s
969:	learn: 0.2004470	total: 20m 49s	remaining: 38.7s
970:	learn: 0.2004061	total: 20m 51s	remaining: 37.4s
971:	learn: 0.2003218	total: 20m 52s	remaining: 36.1s
972:	learn: 0.2002475	total: 20m 53s	remaining: 34.8s
973:	learn: 0.2001728	total: 20m 55s	remaining: 33.5s
974:	learn: 0.2001329	total: 20m 56s	remaining: 32.2s
975:	learn: 0.2000462	total: 20m 57s	remaining: 30.9s
976:	learn: 0.1999553	total: 20m 58s	remaining: 29.6s
977:	learn: 0.1998809	total: 21m	remaining: 28.3s
978:	learn: 0.1998045	total: 21m 1s	remaining: 27.1s
979:	learn: 0.1997648	total: 21m 2s	remaining: 25.8s
980:	learn: 0.1996706	total: 21m 3s	remaining: 24.5s
981:	learn: 0.1995563	total: 21m 5s	r

117:	learn: 0.3719178	total: 2m 39s	remaining: 19m 51s
118:	learn: 0.3714381	total: 2m 40s	remaining: 19m 50s
119:	learn: 0.3710138	total: 2m 42s	remaining: 19m 48s
120:	learn: 0.3704817	total: 2m 43s	remaining: 19m 47s
121:	learn: 0.3695092	total: 2m 44s	remaining: 19m 45s
122:	learn: 0.3690424	total: 2m 46s	remaining: 19m 44s
123:	learn: 0.3685110	total: 2m 47s	remaining: 19m 42s
124:	learn: 0.3681190	total: 2m 48s	remaining: 19m 41s
125:	learn: 0.3676404	total: 2m 50s	remaining: 19m 39s
126:	learn: 0.3669176	total: 2m 51s	remaining: 19m 38s
127:	learn: 0.3665306	total: 2m 52s	remaining: 19m 37s
128:	learn: 0.3659527	total: 2m 54s	remaining: 19m 35s
129:	learn: 0.3651142	total: 2m 55s	remaining: 19m 34s
130:	learn: 0.3643793	total: 2m 56s	remaining: 19m 32s
131:	learn: 0.3639261	total: 2m 58s	remaining: 19m 31s
132:	learn: 0.3633888	total: 2m 59s	remaining: 19m 29s
133:	learn: 0.3625654	total: 3m	remaining: 19m 28s
134:	learn: 0.3620834	total: 3m 2s	remaining: 19m 26s
135:	learn: 0.3

268:	learn: 0.3020919	total: 6m 1s	remaining: 16m 23s
269:	learn: 0.3018038	total: 6m 3s	remaining: 16m 21s
270:	learn: 0.3014109	total: 6m 4s	remaining: 16m 20s
271:	learn: 0.3009584	total: 6m 5s	remaining: 16m 19s
272:	learn: 0.3007255	total: 6m 7s	remaining: 16m 17s
273:	learn: 0.3004075	total: 6m 8s	remaining: 16m 16s
274:	learn: 0.3001724	total: 6m 9s	remaining: 16m 14s
275:	learn: 0.2999565	total: 6m 11s	remaining: 16m 13s
276:	learn: 0.2996543	total: 6m 12s	remaining: 16m 12s
277:	learn: 0.2994254	total: 6m 14s	remaining: 16m 11s
278:	learn: 0.2990810	total: 6m 15s	remaining: 16m 10s
279:	learn: 0.2988449	total: 6m 16s	remaining: 16m 8s
280:	learn: 0.2984647	total: 6m 18s	remaining: 16m 7s
281:	learn: 0.2981986	total: 6m 19s	remaining: 16m 6s
282:	learn: 0.2979308	total: 6m 20s	remaining: 16m 4s
283:	learn: 0.2976760	total: 6m 22s	remaining: 16m 3s
284:	learn: 0.2974472	total: 6m 23s	remaining: 16m 1s
285:	learn: 0.2971651	total: 6m 24s	remaining: 16m
286:	learn: 0.2968745	total

419:	learn: 0.2659072	total: 9m 23s	remaining: 12m 57s
420:	learn: 0.2657900	total: 9m 24s	remaining: 12m 56s
421:	learn: 0.2656165	total: 9m 25s	remaining: 12m 55s
422:	learn: 0.2654811	total: 9m 27s	remaining: 12m 53s
423:	learn: 0.2652710	total: 9m 28s	remaining: 12m 52s
424:	learn: 0.2650842	total: 9m 29s	remaining: 12m 51s
425:	learn: 0.2649716	total: 9m 31s	remaining: 12m 49s
426:	learn: 0.2647422	total: 9m 32s	remaining: 12m 48s
427:	learn: 0.2645043	total: 9m 33s	remaining: 12m 46s
428:	learn: 0.2643356	total: 9m 35s	remaining: 12m 45s
429:	learn: 0.2640610	total: 9m 36s	remaining: 12m 44s
430:	learn: 0.2639073	total: 9m 37s	remaining: 12m 42s
431:	learn: 0.2637254	total: 9m 39s	remaining: 12m 41s
432:	learn: 0.2635704	total: 9m 40s	remaining: 12m 40s
433:	learn: 0.2633873	total: 9m 41s	remaining: 12m 38s
434:	learn: 0.2632873	total: 9m 43s	remaining: 12m 37s
435:	learn: 0.2630026	total: 9m 44s	remaining: 12m 35s
436:	learn: 0.2628330	total: 9m 45s	remaining: 12m 34s
437:	learn

568:	learn: 0.2419428	total: 12m 41s	remaining: 9m 37s
569:	learn: 0.2418010	total: 12m 43s	remaining: 9m 35s
570:	learn: 0.2416042	total: 12m 44s	remaining: 9m 34s
571:	learn: 0.2415419	total: 12m 45s	remaining: 9m 33s
572:	learn: 0.2414298	total: 12m 47s	remaining: 9m 31s
573:	learn: 0.2412238	total: 12m 48s	remaining: 9m 30s
574:	learn: 0.2410880	total: 12m 49s	remaining: 9m 29s
575:	learn: 0.2409555	total: 12m 51s	remaining: 9m 27s
576:	learn: 0.2408215	total: 12m 52s	remaining: 9m 26s
577:	learn: 0.2406763	total: 12m 53s	remaining: 9m 24s
578:	learn: 0.2405361	total: 12m 55s	remaining: 9m 23s
579:	learn: 0.2404555	total: 12m 56s	remaining: 9m 22s
580:	learn: 0.2403286	total: 12m 57s	remaining: 9m 20s
581:	learn: 0.2401378	total: 12m 59s	remaining: 9m 19s
582:	learn: 0.2399945	total: 13m	remaining: 9m 18s
583:	learn: 0.2398690	total: 13m 1s	remaining: 9m 16s
584:	learn: 0.2397406	total: 13m 2s	remaining: 9m 15s
585:	learn: 0.2396073	total: 13m 4s	remaining: 9m 14s
586:	learn: 0.239

718:	learn: 0.2233259	total: 16m	remaining: 6m 15s
719:	learn: 0.2232740	total: 16m 2s	remaining: 6m 14s
720:	learn: 0.2230171	total: 16m 3s	remaining: 6m 12s
721:	learn: 0.2229032	total: 16m 4s	remaining: 6m 11s
722:	learn: 0.2227856	total: 16m 6s	remaining: 6m 10s
723:	learn: 0.2226364	total: 16m 7s	remaining: 6m 8s
724:	learn: 0.2225704	total: 16m 8s	remaining: 6m 7s
725:	learn: 0.2224385	total: 16m 10s	remaining: 6m 6s
726:	learn: 0.2223004	total: 16m 11s	remaining: 6m 4s
727:	learn: 0.2222119	total: 16m 13s	remaining: 6m 3s
728:	learn: 0.2220602	total: 16m 14s	remaining: 6m 2s
729:	learn: 0.2219977	total: 16m 15s	remaining: 6m
730:	learn: 0.2219244	total: 16m 17s	remaining: 5m 59s
731:	learn: 0.2218081	total: 16m 18s	remaining: 5m 58s
732:	learn: 0.2216945	total: 16m 19s	remaining: 5m 56s
733:	learn: 0.2216437	total: 16m 21s	remaining: 5m 55s
734:	learn: 0.2215176	total: 16m 22s	remaining: 5m 54s
735:	learn: 0.2213112	total: 16m 23s	remaining: 5m 52s
736:	learn: 0.2211931	total: 1

869:	learn: 0.2085241	total: 19m 22s	remaining: 2m 53s
870:	learn: 0.2084446	total: 19m 23s	remaining: 2m 52s
871:	learn: 0.2084005	total: 19m 25s	remaining: 2m 51s
872:	learn: 0.2083036	total: 19m 26s	remaining: 2m 49s
873:	learn: 0.2082065	total: 19m 27s	remaining: 2m 48s
874:	learn: 0.2081626	total: 19m 29s	remaining: 2m 47s
875:	learn: 0.2080517	total: 19m 30s	remaining: 2m 45s
876:	learn: 0.2080077	total: 19m 31s	remaining: 2m 44s
877:	learn: 0.2079098	total: 19m 32s	remaining: 2m 42s
878:	learn: 0.2078203	total: 19m 34s	remaining: 2m 41s
879:	learn: 0.2077250	total: 19m 35s	remaining: 2m 40s
880:	learn: 0.2076271	total: 19m 36s	remaining: 2m 38s
881:	learn: 0.2075833	total: 19m 38s	remaining: 2m 37s
882:	learn: 0.2074739	total: 19m 39s	remaining: 2m 36s
883:	learn: 0.2074305	total: 19m 40s	remaining: 2m 34s
884:	learn: 0.2072699	total: 19m 42s	remaining: 2m 33s
885:	learn: 0.2071722	total: 19m 43s	remaining: 2m 32s
886:	learn: 0.2071295	total: 19m 44s	remaining: 2m 30s
887:	learn

### Выводы:

В проекте были обучены модели классификации коментариев на позитивные и негативные, и были получены следующие результаты:
1. Модель логистической регрессии имеет значение метрики F1 = 0.762 на всем наборе данных.
2. Модель CatBoost имеет значение метрики F1 = 0.757 на всем наборе данных.

Для достижения этих показателей классы были взвешены, так как был сильный дисбаланс в целевом признаке.

Обе модели удовлетворяют условиям проекта (F1 должна быть не менее 0.75), но модель логистической регрессии по метрике F1 лучше модели Catboost на 0.5% и скорость обучения модели логистической регрессии выше в 450 (!) раз.

Для решения задачи классификации текстов я бы выбрал модель логистической регрессии.